{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ee429c1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-12-15 15:42:56 Zichens-MacBook-Pro.local metapub.config[59583] WARNING NCBI_API_KEY was not set.\n"
     ]
    }
   ],
   "source": [
    "import argparse\n",
    "from Bio import Entrez\n",
    "\n",
    "import os \n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "import pandas as pd\n",
    "from datetime import date\n",
    "from datetime import datetime\n",
    "import numpy as np\n",
    "import re\n",
    "from metapub import FindIt\n",
    "from metapub.convert import pmid2doi\n",
    "from metapub.convert import doi2pmid\n",
    "from metapub import PubMedFetcher\n",
    "\n",
    "import xml.etree.ElementTree as ET\n",
    "import ssl\n",
    "\n",
    "import ssl\n",
    "ssl._create_default_https_context = ssl._create_unverified_context\n",
    "\n",
    "import argparse\n",
    "from Bio import Entrez\n",
    "\n",
    "ssl._create_default_https_context = ssl._create_unverified_context"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ac51ef2",
   "metadata": {},
   "source": [
    "## Query the NCBI database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fbd0ff62",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use one of the following search queries/filters\n",
    "# filters = \"HiChIP[All Fields] AND \\\"gse\\\"[Filter]\" # the whole notebook will take around 7 min to run\n",
    "# filters = \"HiChIP[All Fields] AND (\\\"gse\\\"[Filter] AND (\\\"Homo sapiens\\\"[Organism] OR \\\"Mus musculus\\\"[Organism]))\n",
    "# filters = \"HiChIP[All Fields] AND (\\\"Homo sapiens\\\"[Organism] AND \\\"gse\\\"[Filter])\"\n",
    "# filters = \"HiChIP[All Fields] AND (\\\"Homo sapiens\\\"[Organism] AND \\\"published last year\\\"[Filter]) AND \\\"gse\\\"[Filter]\" # 2 min\n",
    "# filters = \"HiChIP[All Fields] AND (\\\"Mus musculus\\\"[Organism] AND \\\"gse\\\"[Filter])\"\n",
    "\n",
    "filters = \"HiChIP[All Fields] AND (\\\"gse\\\"[Filter] AND (\\\"Homo sapiens\\\"[Organism] OR \\\"Mus musculus\\\"[Organism]))\"\n",
    "\n",
    "# set a dummy email \n",
    "Entrez.email = \"zjiang@lji.org\"\n",
    "\n",
    "# query the NCBI database \n",
    "search_result = Entrez.esearch(db=\"gds\", retmax=10000, term=filters)\n",
    "result = Entrez.read(search_result)\n",
    "\n",
    "# creating a regex to extract pubmed IDs\n",
    "PubMedIds_pattern = re.compile('[0123456789]+')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d53a4689",
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse the results into a dataframe \n",
    "dictionary_lst = []\n",
    "for Id in result[\"IdList\"]:\n",
    "    handle = Entrez.esummary(db=\"gds\", id=Id, retmode=\"xml\") # get summary of this entry on GEO datasets\n",
    "    entry = Entrez.parse(handle)\n",
    "    for column in entry:\n",
    "        dictionary_lst.append(column)\n",
    "df = pd.DataFrame(dictionary_lst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2e0445ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Item</th>\n",
       "      <th>Id</th>\n",
       "      <th>Accession</th>\n",
       "      <th>GDS</th>\n",
       "      <th>title</th>\n",
       "      <th>summary</th>\n",
       "      <th>GPL</th>\n",
       "      <th>GSE</th>\n",
       "      <th>taxon</th>\n",
       "      <th>entryType</th>\n",
       "      <th>...</th>\n",
       "      <th>ExtRelations</th>\n",
       "      <th>n_samples</th>\n",
       "      <th>SeriesTitle</th>\n",
       "      <th>PlatformTitle</th>\n",
       "      <th>PlatformTaxa</th>\n",
       "      <th>SamplesTaxa</th>\n",
       "      <th>PubMedIds</th>\n",
       "      <th>Projects</th>\n",
       "      <th>FTPLink</th>\n",
       "      <th>GEO2R</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[]</td>\n",
       "      <td>200205218</td>\n",
       "      <td>GSE205218</td>\n",
       "      <td></td>\n",
       "      <td>Auxin-inducible degron 2 system deciphers func...</td>\n",
       "      <td>This SuperSeries is composed of the SubSeries ...</td>\n",
       "      <td>24676;18573</td>\n",
       "      <td>205218</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>54</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE205nn...</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[]</td>\n",
       "      <td>200201911</td>\n",
       "      <td>GSE201911</td>\n",
       "      <td></td>\n",
       "      <td>Auxin-inducible degron 2 system deciphers func...</td>\n",
       "      <td>Background: CTCF is a well-established chromat...</td>\n",
       "      <td>18573</td>\n",
       "      <td>201911</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE201nn...</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[]</td>\n",
       "      <td>200207828</td>\n",
       "      <td>GSE207828</td>\n",
       "      <td></td>\n",
       "      <td>Oct1 recruits the histone lysine demethylase U...</td>\n",
       "      <td>The pathways used by cells to transition betwe...</td>\n",
       "      <td>24247</td>\n",
       "      <td>207828</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE207nn...</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[]</td>\n",
       "      <td>200160943</td>\n",
       "      <td>GSE160943</td>\n",
       "      <td></td>\n",
       "      <td>Oct1 recruits the histone lysine demethylase U...</td>\n",
       "      <td>This SuperSeries is composed of the SubSeries ...</td>\n",
       "      <td>24247</td>\n",
       "      <td>160943</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>32</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE160nn...</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[]</td>\n",
       "      <td>200123645</td>\n",
       "      <td>GSE123645</td>\n",
       "      <td></td>\n",
       "      <td>Super-enhancer driven MYST family histone lysi...</td>\n",
       "      <td>Nasopharyngeal carcinoma (NPC) is causally lin...</td>\n",
       "      <td>15228;18573</td>\n",
       "      <td>123645</td>\n",
       "      <td>synthetic construct; Homo sapiens</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'RelationType': 'SRA', 'TargetObject': 'SRP1...</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE123nn...</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Item         Id  Accession GDS  \\\n",
       "0   []  200205218  GSE205218       \n",
       "1   []  200201911  GSE201911       \n",
       "2   []  200207828  GSE207828       \n",
       "3   []  200160943  GSE160943       \n",
       "4   []  200123645  GSE123645       \n",
       "\n",
       "                                               title  \\\n",
       "0  Auxin-inducible degron 2 system deciphers func...   \n",
       "1  Auxin-inducible degron 2 system deciphers func...   \n",
       "2  Oct1 recruits the histone lysine demethylase U...   \n",
       "3  Oct1 recruits the histone lysine demethylase U...   \n",
       "4  Super-enhancer driven MYST family histone lysi...   \n",
       "\n",
       "                                             summary          GPL     GSE  \\\n",
       "0  This SuperSeries is composed of the SubSeries ...  24676;18573  205218   \n",
       "1  Background: CTCF is a well-established chromat...        18573  201911   \n",
       "2  The pathways used by cells to transition betwe...        24247  207828   \n",
       "3  This SuperSeries is composed of the SubSeries ...        24247  160943   \n",
       "4  Nasopharyngeal carcinoma (NPC) is causally lin...  15228;18573  123645   \n",
       "\n",
       "                               taxon entryType  ...  \\\n",
       "0                       Homo sapiens       GSE  ...   \n",
       "1                       Homo sapiens       GSE  ...   \n",
       "2                       Mus musculus       GSE  ...   \n",
       "3                       Mus musculus       GSE  ...   \n",
       "4  synthetic construct; Homo sapiens       GSE  ...   \n",
       "\n",
       "                                        ExtRelations n_samples SeriesTitle  \\\n",
       "0                                                 []        54               \n",
       "1                                                 []         4               \n",
       "2                                                 []         4               \n",
       "3                                                 []        32               \n",
       "4  [{'RelationType': 'SRA', 'TargetObject': 'SRP1...         9               \n",
       "\n",
       "  PlatformTitle PlatformTaxa SamplesTaxa PubMedIds Projects  \\\n",
       "0                                               []       []   \n",
       "1                                               []       []   \n",
       "2                                               []       []   \n",
       "3                                               []       []   \n",
       "4                                               []       []   \n",
       "\n",
       "                                             FTPLink GEO2R  \n",
       "0  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE205nn...    no  \n",
       "1  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE201nn...    no  \n",
       "2  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE207nn...   yes  \n",
       "3  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE160nn...   yes  \n",
       "4  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE123nn...   yes  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bee10737",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(238, 29)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb4d1e58",
   "metadata": {},
   "source": [
    "## Cleaning the Paper Names and Converting from PMID to DOI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4c5feeac",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "dictionary_lst = []\n",
    "for Id in result[\"IdList\"]:\n",
    "    \n",
    "    handle = Entrez.esummary(db=\"gds\", id=Id, retmode=\"xml\")\n",
    "    entry = Entrez.parse(handle)\n",
    "    \n",
    "    for column in entry:\n",
    "        paper_title = column['title']\n",
    "        new_paper_title = ''\n",
    "        last_word = re.split('\\s|[.]', paper_title)[-1]\n",
    "        last_two_words = re.split('\\s|[.]', paper_title)[-2:]\n",
    "        \n",
    "        # remove \".[HiChIP]\", \". [HiChIP]\", or \"(HiChIP)\" in paper titles using steps below\n",
    "        if ('[' in last_word and ']' in last_word) or \\\n",
    "                ('(' in last_word and ')' in last_word):\n",
    "            new_paper_title = re.split('\\s|[.]', paper_title)[0:-1]\n",
    "            new_paper_title = ' '.join(new_paper_title)\n",
    "            new_paper_title = new_paper_title.strip()\n",
    "            if new_paper_title.endswith('.'): # remove the last period in paper title\n",
    "                new_paper_title = new_paper_title[:-1]\n",
    "                \n",
    "        # remove \".[Hi ChIP]\", \". [Hi ChIP]\", or \"[Bead Array]\" in paper titles using steps below\n",
    "        elif ('[' in last_two_words[0] and ']' in last_two_words[1]) or \\\n",
    "                    ('(' in last_two_words[0] and ')' in last_two_words[1]):\n",
    "            new_paper_title = re.split('\\s|[.]', paper_title)[0:-2]\n",
    "            new_paper_title = ' '.join(new_paper_title)\n",
    "            new_paper_title = new_paper_title.strip()\n",
    "            if new_paper_title.endswith('.'): # remove the last period in paper title\n",
    "                new_paper_title = new_paper_title[:-1]\n",
    "                \n",
    "        elif paper_title.endswith('.'):\n",
    "            new_paper_title = paper_title[:-1] \n",
    "            \n",
    "        else:\n",
    "            new_paper_title = paper_title\n",
    "            \n",
    "        new_paper_title = new_paper_title.replace(\"\\xa0\", \" \") # remove no-break space\n",
    "        column['title'] = new_paper_title\n",
    "        \n",
    "        # convert PMID to DOI, some have no PMID so empty string\n",
    "        string = str(column['PubMedIds'])\n",
    "        pdf_url = \"\"\n",
    "        PubMedIds = PubMedIds_pattern.findall(string)\n",
    "        \n",
    "        if len(PubMedIds) > 0:\n",
    "            DOI = pmid2doi(PubMedIds[0])\n",
    "            column['PubMedIds'] = DOI\n",
    "        else: column['PubMedIds'] = \"\"\n",
    "        dictionary_lst.append(column)\n",
    "        \n",
    "df = pd.DataFrame(dictionary_lst)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c463820d",
   "metadata": {},
   "source": [
    "## Reformat the Columns for Google Sheet Compatibility"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e5bc08ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop extra columns\n",
    "df = pd.DataFrame(dictionary_lst)\n",
    "\n",
    "drop_cols = ['Item', 'Id', 'GDS', 'GPL', 'GSE', 'entryType', 'ptechType',\n",
    "             'valType', 'SSInfo', 'subsetInfo', 'suppFile', 'Relations', 'ExtRelations',\n",
    "             'n_samples', 'SeriesTitle', 'PlatformTitle', 'PlatformTaxa', 'SamplesTaxa',\n",
    "             'Projects', 'FTPLink', 'GEO2R']\n",
    "df.drop(drop_cols, inplace=True, axis=1)\n",
    "\n",
    "# rename columns\n",
    "rename_cols = {\"Accession\":\"GEO / Data link\", \"PubMedIds\":\"DOI\", \"title\":\"Paper Title\",\n",
    "               \"taxon\":\"Organism\", \"gdsType\":\"Any other information\", \"PDAT\":\"Year\",\n",
    "               \"Samples\":\"Other matched data\"}\n",
    "df.rename(columns=rename_cols, inplace=True)\n",
    "\n",
    "# extract just the year\n",
    "df[\"Year\"] = df[\"Year\"].str[:4]\n",
    "\n",
    "# add index for merging\n",
    "df['index'] = np.arange(len(df))\n",
    "\n",
    "# fill na with empty values\n",
    "df.fillna(\"\",inplace=True)\n",
    "\n",
    "# adding missing columns with empty values\n",
    "df[\"Journal\"] = \"\"\n",
    "df[\"Authors\"] = \"\"\n",
    "df[\"Tissue/Cell Line\"] = \"\"\n",
    "df[\"Presenter\"] = \"\"\n",
    "df[\"Potential HiChIP\"] = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19c61b12",
   "metadata": {},
   "source": [
    "## Assigning Potential HiChIP Samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2d5df7f",
   "metadata": {},
   "source": [
    "Filter out samples with no HiChIP mention and assigned Yes or Maybe to all others "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ecfe777c",
   "metadata": {},
   "outputs": [],
   "source": [
    "col = df[\"Other matched data\"]\n",
    "lst = []\n",
    "GSM_IDs = [] ######## new for cell type\n",
    "index = 0\n",
    "for row in col:\n",
    "    \n",
    "    temp = []\n",
    "    state = False\n",
    "    \n",
    "    # Check for the term HiChIP in the title\n",
    "    for ele in row: \n",
    "        if (\"HiChIP\".casefold() in ele[\"Title\"].casefold()) or \\\n",
    "                (\"Hi-ChIP\".casefold() in ele[\"Title\"].casefold()):\n",
    "            state = True\n",
    "    \n",
    "    # keep samples with HiChIP and mark \"Yes\"\n",
    "    if state == True: \n",
    "        for ele in row:\n",
    "            GSM_IDs.append(ele[\"Accession\"]) ######## new for cell type\n",
    "            if (\"HiChIP\".casefold() in ele[\"Title\"].casefold()) or \\\n",
    "                    (\"Hi-ChIP\".casefold() in ele[\"Title\"].casefold()):\n",
    "                temp.append(ele[\"Accession\"] + \": \" + ele[\"Title\"])\n",
    "        temp_str = \"\\n\".join(temp)\n",
    "        df.at[index, \"Potential HiChIP\"] = \"Yes\"\n",
    "\n",
    "    # otherwise, keep all GSM samples and mark \"Maybe\"\n",
    "    else: \n",
    "        for ele in row:\n",
    "            GSM_IDs.append(ele[\"Accession\"]) ######## new for cell type\n",
    "            temp.append(ele[\"Accession\"] + \": \" + ele[\"Title\"])\n",
    "        temp_str = \"\\n\".join(temp)\n",
    "        df.at[index,\"Potential HiChIP\"]=\"Maybe\"\n",
    "    lst.append(temp_str)\n",
    "    index += 1\n",
    "    \n",
    "df[\"Other matched data\"] = list(lst)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e88c1586",
   "metadata": {},
   "source": [
    "## Merge Rows with the Same Paper Title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4aa57284",
   "metadata": {},
   "outputs": [],
   "source": [
    "aggregation_functions = {'Paper Title':'first', 'DOI':'max', 'Journal':'first', 'Authors':'first', 'Year':'first', 'GEO / Data link':lambda x: '\\n'.join(x),\n",
    "       'Any other information':lambda x: '\\n'.join(x), 'Organism':'first', 'Tissue/Cell Line':'first',\n",
    "       'Potential HiChIP':'first', 'Other matched data':lambda x: '\\n'.join(x), 'Presenter':'first'}\n",
    "df_grouped = df.groupby(df['Paper Title']).aggregate(aggregation_functions)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a13db4f0",
   "metadata": {},
   "source": [
    "## Fetch Journal Name and First Author of Each Paper and Add https to DOI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55a1a9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "jounrnal_lst = []\n",
    "authors_lst = []\n",
    "fetch = PubMedFetcher()\n",
    "for DOI in df_grouped['DOI']:\n",
    "    try:\n",
    "        PMID = doi2pmid(DOI)\n",
    "        article = fetch.article_by_pmid(PMID)\n",
    "        jounrnal_lst.append(article.journal)\n",
    "        authors_lst.append(article.authors[0].split()[0] + \" et al.\")\n",
    "    except:\n",
    "        jounrnal_lst.append(\"\")\n",
    "        authors_lst.append(\"\")\n",
    "df_grouped['Journal'] = jounrnal_lst\n",
    "df_grouped['Authors'] = authors_lst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9384d88e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add https address to DOI (for easy access)\n",
    "df_grouped.loc[(df_grouped['DOI'] != ''), 'DOI'] = 'https://doi.org/' + df_grouped.loc[(df_grouped['DOI'] != ''), 'DOI']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07a4b4c8",
   "metadata": {},
   "source": [
    "## Note the Date Added, Set Output Path, and Save the Final Output File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beed3d3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make an output directory\n",
    "######## new for cell type\n",
    "outdir = ''\n",
    "# outdir = 'results/hichip_db/'\n",
    "# os.makedirs(outdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be627b76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# determining the current year, month and day\n",
    "today = date.today()\n",
    "date_str = today.strftime(\"%Y_%m_%d\")\n",
    "\n",
    "# determining current hour and minute\n",
    "now = datetime.now()\n",
    "time_str = now.strftime(\"%H_%M\")\n",
    "\n",
    "# setting the output filename\n",
    "output = os.path.join(outdir, \"GEO_Query.{}_{}\".format(date_str, time_str))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1e2e626",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add a column that notes the added date\n",
    "df_grouped['Date Added'] = date_str.replace('_', '-')\n",
    "reorder = ['Paper Title', 'Journal', 'Authors', 'Year', 'DOI', 'GEO / Data link', 'Any other information',\n",
    "           'Organism', 'Tissue/Cell Line', 'Potential HiChIP', 'Other matched data', 'Presenter', 'Date Added']\n",
    "df_grouped = df_grouped.loc[:, reorder]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d90d666",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_grouped.to_excel(output+\".xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "751f480e",
   "metadata": {},
   "source": [
    "## Save the GSM IDs line by line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3de96405",
   "metadata": {},
   "outputs": [],
   "source": [
    "GSM_IDs = list(set(GSM_IDs))\n",
    "with open(r'gsm_list.txt', 'w') as fp:\n",
    "    for gsm in GSM_IDs:\n",
    "        # write each item on a new line\n",
    "        fp.write(\"%s\\n\" % gsm)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a0186bb",
   "metadata": {},
   "source": [
    "## Query GEO BioSample using the GSM IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49d34276",
   "metadata": {},
   "outputs": [],
   "source": [
    "GSM_IDs_from_file = []\n",
    "# open file and read the content in a list\n",
    "with open(r'gsm_list.txt', 'r') as fp:\n",
    "    for line in fp:\n",
    "        # remove linebreak from a current name\n",
    "        # linebreak is the last character of each line\n",
    "        gsm = line[:-1]\n",
    "\n",
    "        # add current item to the list\n",
    "        GSM_IDs_from_file.append(gsm)\n",
    "\n",
    "GSM_filters = ' [All Fields] OR '.join(GSM_IDs_from_file[0:1000])\n",
    "GSM_filters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6794baee",
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = \"zjiang@lji.org\"\n",
    "\n",
    "search_result = Entrez.esearch(db=\"biosample\", retmax=10000, term=GSM_filters)\n",
    "result = Entrez.read(search_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bbdd07e",
   "metadata": {},
   "source": [
    "### Filter query result by looping through XML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d24f12b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_df = pd.DataFrame(columns=[\"GEO\", 'organism', 'biomaterial', 'tissue', 'celltype', 'strain', 'age', 'genotype', 'treatment', 'antibody', 'other'])\n",
    "\n",
    "for Id in result[\"IdList\"]:\n",
    "    handle = Entrez.esummary(db=\"biosample\", id=Id, retmode=\"xml\") # get summary of this entry on GEO datasets\n",
    "    entry = Entrez.read(handle)\n",
    "    record = entry[\"DocumentSummarySet\"][\"DocumentSummary\"][0][\"SampleData\"]\n",
    "    myroot = ET.fromstring(record)\n",
    "    gsm_data = {\n",
    "        \"GEO\":\"\",\n",
    "        \"organism\":\"\",\n",
    "        \"biomaterial\":\"\",\n",
    "        \"tissue\":\"\",\n",
    "        \"celltype\":\"\",\n",
    "        \"strain\":\"\",\n",
    "        \"age\":\"\",\n",
    "        \"genotype\":\"\",\n",
    "        \"treatment\":\"\",\n",
    "        \"antibody\":\"\",\n",
    "        \"other\":[]\n",
    "        }\n",
    "\n",
    "    for elem in myroot:\n",
    "        for subelem in elem:\n",
    "            if ('db\\': \\'GEO' in str(subelem.attrib)):\n",
    "                gsm_data[\"GEO\"] = subelem.text\n",
    "            if \"taxonomy_name\" in str(subelem.attrib):\n",
    "                gsm_data[\"organism\"] = subelem.attrib[\"taxonomy_name\"]\n",
    "            if ('source_name' in str(subelem.attrib)):\n",
    "                gsm_data[\"biomaterial\"] = subelem.text\n",
    "            if ('cell_type' in str(subelem.attrib)):\n",
    "                gsm_data[\"tissue\"] = subelem.text\n",
    "            if ('cell_line' in str(subelem.attrib)):\n",
    "                gsm_data[\"celltype\"] = subelem.text\n",
    "            if ('strain' in str(subelem.attrib)):\n",
    "                gsm_data[\"strain\"] = subelem.text\n",
    "            if ('age' in str(subelem.attrib)):\n",
    "                gsm_data[\"age\"] = subelem.text\n",
    "            if ('genotype' in str(subelem.attrib)):\n",
    "                gsm_data[\"genotype\"] = subelem.text\n",
    "            if ('treatment' in str(subelem.attrib)):\n",
    "                gsm_data[\"treatment\"] = subelem.text\n",
    "            if ('antibody' in str(subelem.attrib)):\n",
    "                gsm_data[\"antibody\"] = subelem.text\n",
    "            if (len(str(subelem.attrib)) > 2 and \\\n",
    "                'url' not in str(subelem.attrib) and \\\n",
    "                'BioSample' not in str(subelem.attrib) and \\\n",
    "                'SRA' not in str(subelem.attrib) and \\\n",
    "                'bioproject' not in str(subelem.attrib) and \\\n",
    "                'db\\': \\'GEO' not in str(subelem.attrib) and \\\n",
    "                'taxonomy_name' not in str(subelem.attrib) and \\\n",
    "                'source_name' not in str(subelem.attrib) and \\\n",
    "                'cell_type' not in str(subelem.attrib) and \\\n",
    "                'cell_line' not in str(subelem.attrib) and \\\n",
    "                'strain' not in str(subelem.attrib) and \\\n",
    "                'age' not in str(subelem.attrib) and \\\n",
    "                'genotype' not in str(subelem.attrib) and \\\n",
    "                'treatment' not in str(subelem.attrib) and \\\n",
    "                'antibody' not in str(subelem.attrib)\n",
    "               ):\n",
    "                gsm_data[\"other\"].append(str(subelem.attrib) + \"-->\" + subelem.text)\n",
    "    gsm_data[\"other\"] = '|'.join(gsm_data[\"other\"])\n",
    "    cell_type_df = cell_type_df.append(gsm_data, ignore_index=True)\n",
    "    \n",
    "    \n",
    "# \tGSM4027555 and more\n",
    "#     <Attribute attribute_name=\"type\">Tumor</Attribute>\n",
    "# \t\t<Attribute attribute_name=\"relapse\">No</Attribute>\n",
    "# \t\t<Attribute attribute_name=\"gender\" harmonized_name=\"sex\" display_name=\"sex\">male</Attribute>\n",
    "# \t\t<Attribute attribute_name=\"age_days\">2338</Attribute> or timepoint\n",
    "# \t\t<Attribute attribute_name=\"Stage\">4</Attribute>\n",
    "\n",
    "# GSM6528200 and more\n",
    "# <Attribute attribute_name=\"disease state\" harmonized_name=\"disease\" display_name=\"disease\">HCC</Attribute>\n",
    "# <Attribute attribute_name=\"tissue\" harmonized_name=\"tissue\" display_name=\"tissue\">tumour</Attribute>\n",
    "\n",
    "# GSM3930313 and more\n",
    "# {'attribute_name': 'biological sample'}-->FHC should replace celltype"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d75fe44",
   "metadata": {},
   "source": [
    "### Filter query result using root.find()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b788dfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# root.find(\"./Ids/Id[@db='GEO']\" too specific, some attributes have a slightly different synynom as name\n",
    "# cell_type_df = pd.DataFrame(columns=[\"GEO\", 'organism', 'biomaterial', 'tissue', 'celltype', 'strain', 'age', 'genotype', 'treatment', 'antibody', 'other'])\n",
    "\n",
    "# for Id in result[\"IdList\"]:\n",
    "#     handle = Entrez.esummary(db=\"biosample\", id=Id, retmode=\"xml\") # get summary of this entry on GEO datasets\n",
    "#     entry = Entrez.read(handle)\n",
    "#     record = entry[\"DocumentSummarySet\"][\"DocumentSummary\"][0][\"SampleData\"]\n",
    "#     root = ET.fromstring(record)\n",
    "#     gsm_data = {\n",
    "#         \"GEO\":\"\",\n",
    "#         \"organism\":\"\",\n",
    "#         \"biomaterial\":\"\",\n",
    "#         \"tissue\":\"\",\n",
    "#         \"celltype\":\"\",\n",
    "#         \"strain\":\"\",\n",
    "#         \"age\":\"\",\n",
    "#         \"genotype\":\"\",\n",
    "#         \"treatment\":\"\",\n",
    "#         \"antibody\":\"\",\n",
    "#         \"other\":[]\n",
    "#         }\n",
    "    \n",
    "#     try:\n",
    "#         gsm_data[\"GEO\"] = root.find(\"./Ids/Id[@db='GEO']\").text\n",
    "#     except:\n",
    "#         gsm_data[\"GEO\"] = \"\"\n",
    "        \n",
    "        \n",
    "#     try:\n",
    "#         gsm_data[\"organism\"] = root.find(\"./Description/Organism/OrganismName\").text\n",
    "#     except:\n",
    "#         gsm_data[\"organism\"] = \"\"\n",
    "        \n",
    "#     try:\n",
    "#         gsm_data[\"biomaterial\"] = root.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text\n",
    "#     except:\n",
    "#         gsm_data[\"biomaterial\"] = \"\"\n",
    "        \n",
    "#     try:\n",
    "#         gsm_data[\"tissue\"] = root.find(\"./Attributes/Attribute[@attribute_name='cell type']\").text\n",
    "#     except:\n",
    "#         gsm_data[\"tissue\"] = \"\"\n",
    "        \n",
    "        \n",
    "#     try:\n",
    "#         gsm_data[\"celltype\"] = root.find(\"./Attributes/Attribute[@attribute_name='cell line']\").text\n",
    "#     except:\n",
    "#         gsm_data[\"celltype\"] = \"\"\n",
    "        \n",
    "#     try:\n",
    "#         gsm_data[\"strain\"] = root.find(\"./Attributes/Attribute[@attribute_name='strain']\").text\n",
    "#     except:\n",
    "#         gsm_data[\"strain\"] = \"\"\n",
    "        \n",
    "#     try:\n",
    "#         gsm_data[\"age\"] = root.find(\"./Attributes/Attribute[@attribute_name='age']\").text\n",
    "#     except:\n",
    "#         gsm_data[\"age\"] = \"\"\n",
    "        \n",
    "#     try:\n",
    "#         gsm_data[\"genotype\"] = root.find(\"./Attributes/Attribute[@attribute_name='genotype']\").text\n",
    "#     except:\n",
    "#         gsm_data[\"genotype\"] = \"\"\n",
    "        \n",
    "#     try:\n",
    "#         gsm_data[\"treatment\"] = root.find(\"./Attributes/Attribute[@attribute_name='treatment']\").text\n",
    "#     except:\n",
    "#         gsm_data[\"treatment\"] = \"\"\n",
    "        \n",
    "#     try:\n",
    "#         gsm_data[\"antibody\"] = root.find(\"./Attributes/Attribute[@attribute_name='chip antibody']\").text\n",
    "#     except:\n",
    "#         gsm_data[\"antibody\"] = \"\"\n",
    "    \n",
    "#     for elem in myroot:\n",
    "#         for subelem in elem:\n",
    "#             if (len(str(subelem.attrib)) > 2 and \\\n",
    "#                 'url' not in str(subelem.attrib) and \\\n",
    "#                 'BioSample' not in str(subelem.attrib) and \\\n",
    "#                 'SRA' not in str(subelem.attrib) and \\\n",
    "#                 'bioproject' not in str(subelem.attrib) and \\\n",
    "#                 'db\\': \\'GEO' not in str(subelem.attrib) and \\\n",
    "#                 'taxonomy_name' not in str(subelem.attrib) and \\\n",
    "#                 'source_name' not in str(subelem.attrib) and \\\n",
    "#                 'cell_type' not in str(subelem.attrib) and \\\n",
    "#                 'cell_line' not in str(subelem.attrib) and \\\n",
    "#                 'strain' not in str(subelem.attrib) and \\\n",
    "#                 'age' not in str(subelem.attrib) and \\\n",
    "#                 'genotype' not in str(subelem.attrib) and \\\n",
    "#                 'treatment' not in str(subelem.attrib) and \\\n",
    "#                 'antibody' not in str(subelem.attrib)\n",
    "#                ):\n",
    "#                 gsm_data[\"other\"].append(str(subelem.attrib) + \"-->\" + subelem.text)\n",
    "#     gsm_data[\"other\"] = '|'.join(gsm_data[\"other\"])\n",
    "    \n",
    "#     cell_type_df = cell_type_df.append(gsm_data, ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "646b82b9",
   "metadata": {},
   "source": [
    "### Save query result filtered table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c37ab216",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_df.to_excel(\"test2.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5a70117",
   "metadata": {},
   "source": [
    "Sample ID: This is the unique identifier for each sample in the database. You can search for samples by their ID using this field.\n",
    "\n",
    "Organism: This field allows you to search for samples from a specific organism, such as human, mouse, or yeast.\n",
    "\n",
    "Tissue: This field allows you to search for samples based on their tissue type, such as liver, brain, or heart.\n",
    "\n",
    "Disease state: This field allows you to search for samples based on their disease state, such as cancer, diabetes, or autoimmune disease.\n",
    "\n",
    "Cell type: This field allows you to search for samples based on their cell type, such as fibroblasts, macrophages, or stem cells.\n",
    "\n",
    "Developmental stage: This field allows you to search for samples based on their developmental stage, such as embryonic, fetal, or adult.\n",
    "\n",
    "Experimental design: This field allows you to search for samples based on the type of experiment they were used in, such as microarray, RNA-seq, or ChIP-seq."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9267f94",
   "metadata": {},
   "source": [
    "### GEO BioSample query result for one GSM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c473797b",
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = \"your_email@example.com\"  # Provide your email address\n",
    "\n",
    "# Retrieve all the BioSample records\n",
    "handle = Entrez.esearch(db=\"biosample\", term=\"GSM5029770\")\n",
    "records = Entrez.read(handle)\n",
    "\n",
    "# Iterate through all the records and print the fields and values\n",
    "for record in records[\"IdList\"]:\n",
    "    handle = Entrez.esummary(db=\"biosample\", id=record)\n",
    "    record = Entrez.read(handle)\n",
    "    data = record[\"DocumentSummarySet\"][\"DocumentSummary\"][0][\"SampleData\"]\n",
    "    print(data, \"\\n\")\n",
    "    myroot = ET.fromstring(data)\n",
    "    for elem in myroot:\n",
    "        for subelem in elem:\n",
    "            if \"taxonomy_name\" in str(subelem.attrib):\n",
    "                print(subelem.attrib, \"-->\", subelem.attrib[\"taxonomy_name\"])\n",
    "            else:\n",
    "                print(subelem.attrib, \"-->\", subelem.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90ce0600",
   "metadata": {},
   "source": [
    "### Dict of organs and their synynoms for classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a298f6a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# organ names are based on ENCODE > data > experiment matrix and common organ names returned by ChatGPT\n",
    "dic = {\n",
    "    \"blood\": [\"vessel\"],\n",
    "    \"blodily fluid\": [],\n",
    "    \"epithelium\": [\"skin\", \"epi\"],\n",
    "    \"brain\": [],\n",
    "    \"endocrine gland\": [\"thyroid gland\"],\n",
    "    \"exocrine gland\": [\"lymph node\"],\n",
    "    \"liver\": [],\n",
    "    \"heart\": [],\n",
    "    \"embryo\": [],\n",
    "    \"connective tissue\": [],\n",
    "    \"musculature of body\": [\"muscul\"],\n",
    "    \"lung\": [],\n",
    "    \"intestine\": [],\n",
    "    \"kidney\": [],\n",
    "    \"bone marrow\": [],\n",
    "    \"colon\": [],\n",
    "    \"spleen\": [],\n",
    "    \"stomach\": [],\n",
    "    \"placenta\": [],\n",
    "    \"pancreas\": [],\n",
    "    \"bladder\": [],\n",
    "    \"reproductive organs\": [\"uterus\", \"penis\", \"ovar\", \"test\"],\n",
    "}\n",
    "# nerves?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0509dd21",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fed0c88e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "838f76b8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
