{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ee429c1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-02-19 18:49:05 herman-login3.hpc.lji.org metapub.config[16877] WARNING NCBI_API_KEY was not set.\n"
     ]
    }
   ],
   "source": [
    "import os \n",
    "import re\n",
    "import requests\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "import ssl\n",
    "ssl._create_default_https_context = ssl._create_unverified_context\n",
    "\n",
    "from datetime import date\n",
    "from datetime import datetime\n",
    "\n",
    "from metapub.convert import doi2pmid\n",
    "from metapub import PubMedFetcher\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from lxml import etree\n",
    "from Bio import Entrez"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ac51ef2",
   "metadata": {},
   "source": [
    "## Query the NCBI database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fbd0ff62",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use one of the following search queries/filters\n",
    "# filters = \"HiChIP[All Fields] AND \\\"gse\\\"[Filter]\" # the whole notebook will take around 7 min to run\n",
    "# filters = \"HiChIP[All Fields] AND (\\\"gse\\\"[Filter] AND (\\\"Homo sapiens\\\"[Organism] OR \\\"Mus musculus\\\"[Organism]))\n",
    "# filters = \"HiChIP[All Fields] AND (\\\"Homo sapiens\\\"[Organism] AND \\\"gse\\\"[Filter])\"\n",
    "# filters = \"HiChIP[All Fields] AND (\\\"Homo sapiens\\\"[Organism] AND \\\"published last year\\\"[Filter]) AND \\\"gse\\\"[Filter]\" # 2 min\n",
    "# filters = \"HiChIP[All Fields] AND (\\\"Mus musculus\\\"[Organism] AND \\\"gse\\\"[Filter])\"\n",
    "\n",
    "# set a dummy email \n",
    "Entrez.email = \"zjiang@lji.org\"\n",
    "\n",
    "# query the NCBI database \n",
    "filters = \"HiChIP[All Fields] AND (\\\"gse\\\"[Filter] AND (\\\"Homo sapiens\\\"[Organism] OR \\\"Mus musculus\\\"[Organism]))\"\n",
    "search_result = Entrez.esearch(db=\"gds\", retmax=10000, term=filters)\n",
    "ncbi_results = Entrez.read(search_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5cb91ecf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get summary of this entry on GEO datasets (abbreviated as GDS)\n",
    "ncbi_ids = ','.join(ncbi_results['IdList'])\n",
    "geo_query_handle = Entrez.esummary(db=\"gds\", id=ncbi_ids, retmode=\"xml\") \n",
    "geo_entries = list(Entrez.parse(geo_query_handle))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb4d1e58",
   "metadata": {},
   "source": [
    "## Cleaning the Paper Names and Converting from PMID to DOI"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd1684e3",
   "metadata": {},
   "source": [
    "Extract PMIDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eb8369c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['27643841',\n",
       " '28945252',\n",
       " '29224777',\n",
       " '29641996',\n",
       " '29706538',\n",
       " '29731168',\n",
       " '30046115',\n",
       " '30122536',\n",
       " '30397335',\n",
       " '30580963']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pubmed_ids = set()\n",
    "for entry in geo_entries:\n",
    "    for value in entry['PubMedIds']:\n",
    "        pubmed_ids.add(str(int(value)))\n",
    "pubmed_ids = sorted(pubmed_ids)\n",
    "pubmed_ids[0:10]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "975347e7",
   "metadata": {},
   "source": [
    "Convert from PMID to DOI using the eutils API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "63673aca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the base URL for the efetch command\n",
    "efetch_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi'\n",
    "\n",
    "# define the parameters for the efetch command\n",
    "params = {\n",
    "    'db': 'pubmed',\n",
    "    'id': ','.join(pubmed_ids),\n",
    "    'retmode': 'xml',\n",
    "    'rettype': 'docsum'\n",
    "}\n",
    "\n",
    "# send the request to the E-utilities API and get the response\n",
    "ds_response = requests.get(efetch_url, params=params)\n",
    "\n",
    "# parse the XML response and extract the DOIs\n",
    "ds_root = etree.fromstring(ds_response.content)\n",
    "doc_sums = ds_root.findall('DocSum')\n",
    "pubmed_data = []\n",
    "for doc_sum in doc_sums:\n",
    "\n",
    "    pmid = doc_sum.find('Id').text\n",
    "    dois = [x.text for x in doc_sum.findall('Item[@Name=\"DOI\"]')]\n",
    "    dois = ','.join(dois)\n",
    "    pubmed_data.append([pmid, dois])\n",
    "\n",
    "pubmed_data = pd.DataFrame(pubmed_data, columns=['PMID', 'DOI'])\n",
    "pubmed_data.set_index('PMID', inplace=True)\n",
    "pubmed_data = pubmed_data.squeeze()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "5f1bbb57",
   "metadata": {},
   "source": [
    "#### Convert GEO entries into a dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4c5feeac",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "geo_data = []\n",
    "for i, entry in enumerate(geo_entries):\n",
    "\n",
    "    ##### Cleaning up the title #####\n",
    "    paper_title = entry['title']\n",
    "    new_paper_title = ''\n",
    "    last_word = re.split('\\s|[.]', paper_title)[-1]\n",
    "    last_two_words = re.split('\\s|[.]', paper_title)[-2:]\n",
    "    \n",
    "    # remove \".[HiChIP]\", \". [HiChIP]\", or \"(HiChIP)\" in paper titles using steps below\n",
    "    if ('[' in last_word and ']' in last_word) or \\\n",
    "            ('(' in last_word and ')' in last_word):\n",
    "        new_paper_title = re.split('\\s|[.]', paper_title)[0:-1]\n",
    "        new_paper_title = ' '.join(new_paper_title)\n",
    "        new_paper_title = new_paper_title.strip()\n",
    "\n",
    "        # remove the last period in paper title\n",
    "        if new_paper_title.endswith('.'):\n",
    "            new_paper_title = new_paper_title[:-1]\n",
    "            \n",
    "    # remove \".[Hi ChIP]\", \". [Hi ChIP]\", or \"[Bead Array]\" in paper titles using steps below\n",
    "    elif ('[' in last_two_words[0] and ']' in last_two_words[1]) or \\\n",
    "                ('(' in last_two_words[0] and ')' in last_two_words[1]):\n",
    "        new_paper_title = re.split('\\s|[.]', paper_title)[0:-2]\n",
    "        new_paper_title = ' '.join(new_paper_title)\n",
    "        new_paper_title = new_paper_title.strip()\n",
    "\n",
    "        # remove the last period in paper title\n",
    "        if new_paper_title.endswith('.'):\n",
    "            new_paper_title = new_paper_title[:-1]\n",
    "            \n",
    "    # remove the period\n",
    "    elif paper_title.endswith('.'):\n",
    "        new_paper_title = paper_title[:-1] \n",
    "        \n",
    "    else:\n",
    "        new_paper_title = paper_title\n",
    "\n",
    "    new_paper_title = new_paper_title.replace(\"\\xa0\", \" \") # remove no-break space\n",
    "    entry['title'] = new_paper_title\n",
    "\n",
    "    ##### add the doi to the entry #####\n",
    "    curr_dois = []\n",
    "    for value in entry['PubMedIds']:\n",
    "        pubmed_id = str(int(value))\n",
    "        curr_doi = pubmed_data[pubmed_id]\n",
    "        curr_dois.append(curr_doi)\n",
    "    entry['DOI'] = ','.join(curr_dois)\n",
    "\n",
    "    ##### add the adjusted entry to the data #####\n",
    "    geo_data.append(entry)\n",
    "    \n",
    "geo_df = pd.DataFrame(geo_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b7fbf94b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Item</th>\n",
       "      <th>Id</th>\n",
       "      <th>Accession</th>\n",
       "      <th>GDS</th>\n",
       "      <th>title</th>\n",
       "      <th>summary</th>\n",
       "      <th>GPL</th>\n",
       "      <th>GSE</th>\n",
       "      <th>taxon</th>\n",
       "      <th>entryType</th>\n",
       "      <th>...</th>\n",
       "      <th>n_samples</th>\n",
       "      <th>SeriesTitle</th>\n",
       "      <th>PlatformTitle</th>\n",
       "      <th>PlatformTaxa</th>\n",
       "      <th>SamplesTaxa</th>\n",
       "      <th>PubMedIds</th>\n",
       "      <th>Projects</th>\n",
       "      <th>FTPLink</th>\n",
       "      <th>GEO2R</th>\n",
       "      <th>DOI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[]</td>\n",
       "      <td>200223259</td>\n",
       "      <td>GSE223259</td>\n",
       "      <td></td>\n",
       "      <td>Chromatin accessibilities in C9HRE ALS/FTD pat...</td>\n",
       "      <td>This SuperSeries is composed of the SubSeries ...</td>\n",
       "      <td>16791</td>\n",
       "      <td>223259</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>24</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE223nn...</td>\n",
       "      <td>yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[]</td>\n",
       "      <td>200223258</td>\n",
       "      <td>GSE223258</td>\n",
       "      <td></td>\n",
       "      <td>To profile the chromatin structural changes re...</td>\n",
       "      <td>DAXX is an epigenetic regulator and chromatin ...</td>\n",
       "      <td>16791</td>\n",
       "      <td>223258</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE223nn...</td>\n",
       "      <td>yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[]</td>\n",
       "      <td>200178673</td>\n",
       "      <td>GSE178673</td>\n",
       "      <td></td>\n",
       "      <td>Chromatin landscape in perinatal cardiomyocytes</td>\n",
       "      <td>In order to understand the transcriptional reg...</td>\n",
       "      <td>21273;24688</td>\n",
       "      <td>178673</td>\n",
       "      <td>Mus musculus; Rattus norvegicus</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[IntegerElement(36653336, attributes={})]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE178nn...</td>\n",
       "      <td>yes</td>\n",
       "      <td>10.1038/s41420-023-01322-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[]</td>\n",
       "      <td>200215020</td>\n",
       "      <td>GSE215020</td>\n",
       "      <td></td>\n",
       "      <td>Generate 3D genome database for atrial and ven...</td>\n",
       "      <td>Measurement the contribution of 3D genome stru...</td>\n",
       "      <td>24247</td>\n",
       "      <td>215020</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE215nn...</td>\n",
       "      <td>yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[]</td>\n",
       "      <td>200215065</td>\n",
       "      <td>GSE215065</td>\n",
       "      <td></td>\n",
       "      <td>Regulation of mouse chamber selective enhancers</td>\n",
       "      <td>This SuperSeries is composed of the SubSeries ...</td>\n",
       "      <td>19057;26526;24247</td>\n",
       "      <td>215065</td>\n",
       "      <td>synthetic construct; Mus musculus</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>82</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[IntegerElement(36705030, attributes={})]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE215nn...</td>\n",
       "      <td>yes</td>\n",
       "      <td>10.1161/CIRCULATIONAHA.122.061955</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Item         Id  Accession GDS  \\\n",
       "0   []  200223259  GSE223259       \n",
       "1   []  200223258  GSE223258       \n",
       "2   []  200178673  GSE178673       \n",
       "3   []  200215020  GSE215020       \n",
       "4   []  200215065  GSE215065       \n",
       "\n",
       "                                               title  \\\n",
       "0  Chromatin accessibilities in C9HRE ALS/FTD pat...   \n",
       "1  To profile the chromatin structural changes re...   \n",
       "2    Chromatin landscape in perinatal cardiomyocytes   \n",
       "3  Generate 3D genome database for atrial and ven...   \n",
       "4    Regulation of mouse chamber selective enhancers   \n",
       "\n",
       "                                             summary                GPL  \\\n",
       "0  This SuperSeries is composed of the SubSeries ...              16791   \n",
       "1  DAXX is an epigenetic regulator and chromatin ...              16791   \n",
       "2  In order to understand the transcriptional reg...        21273;24688   \n",
       "3  Measurement the contribution of 3D genome stru...              24247   \n",
       "4  This SuperSeries is composed of the SubSeries ...  19057;26526;24247   \n",
       "\n",
       "      GSE                              taxon entryType  ... n_samples  \\\n",
       "0  223259                       Homo sapiens       GSE  ...        24   \n",
       "1  223258                       Homo sapiens       GSE  ...         3   \n",
       "2  178673    Mus musculus; Rattus norvegicus       GSE  ...        28   \n",
       "3  215020                       Mus musculus       GSE  ...         6   \n",
       "4  215065  synthetic construct; Mus musculus       GSE  ...        82   \n",
       "\n",
       "  SeriesTitle PlatformTitle PlatformTaxa SamplesTaxa  \\\n",
       "0                                                      \n",
       "1                                                      \n",
       "2                                                      \n",
       "3                                                      \n",
       "4                                                      \n",
       "\n",
       "                                   PubMedIds Projects  \\\n",
       "0                                         []       []   \n",
       "1                                         []       []   \n",
       "2  [IntegerElement(36653336, attributes={})]       []   \n",
       "3                                         []       []   \n",
       "4  [IntegerElement(36705030, attributes={})]       []   \n",
       "\n",
       "                                             FTPLink GEO2R  \\\n",
       "0  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE223nn...   yes   \n",
       "1  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE223nn...   yes   \n",
       "2  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE178nn...   yes   \n",
       "3  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE215nn...   yes   \n",
       "4  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE215nn...   yes   \n",
       "\n",
       "                                 DOI  \n",
       "0                                     \n",
       "1                                     \n",
       "2         10.1038/s41420-023-01322-3  \n",
       "3                                     \n",
       "4  10.1161/CIRCULATIONAHA.122.061955  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geo_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c463820d",
   "metadata": {},
   "source": [
    "## Reformat the Columns for Google Sheet Compatibility"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e5bc08ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop columns\n",
    "drop_cols = ['Item', 'Id', 'GDS', 'GPL', 'GSE', 'entryType', 'ptechType',\n",
    "             'valType', 'SSInfo', 'subsetInfo', 'suppFile', 'Relations', 'ExtRelations',\n",
    "             'n_samples', 'SeriesTitle', 'PlatformTitle', 'PlatformTaxa', 'SamplesTaxa',\n",
    "             'Projects', 'FTPLink', 'GEO2R']\n",
    "geo_df.drop(drop_cols, inplace=True, axis=1)\n",
    "\n",
    "# rename columns\n",
    "rename_cols = {\"Accession\":\"GEO / Data link\",\n",
    "               \"title\": \"Paper Title\",\n",
    "               \"taxon\": \"Organism\",\n",
    "               \"gdsType\": \"Any other information\",\n",
    "               \"PDAT\": \"Year\",\n",
    "               \"Samples\": \"Other matched data\"}\n",
    "geo_df.rename(columns=rename_cols, inplace=True)\n",
    "\n",
    "# extract just the year\n",
    "geo_df[\"Year\"] = geo_df[\"Year\"].str[:4]\n",
    "\n",
    "# add index for merging\n",
    "geo_df['index'] = np.arange(len(geo_df))\n",
    "\n",
    "# fill na with empty values\n",
    "geo_df.fillna(\"\", inplace=True)\n",
    "\n",
    "# adding missing columns with empty values\n",
    "geo_df[\"Journal\"] = \"\"\n",
    "geo_df[\"Authors\"] = \"\"\n",
    "geo_df[\"Tissue/Cell Line\"] = \"\"\n",
    "geo_df[\"Presenter\"] = \"\"\n",
    "geo_df[\"Potential HiChIP\"] = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19c61b12",
   "metadata": {},
   "source": [
    "## Assigning Potential HiChIP Samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2d5df7f",
   "metadata": {},
   "source": [
    "Filter out samples with no HiChIP mention and assigned Yes or Maybe to all others "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ecfe777c",
   "metadata": {},
   "outputs": [],
   "source": [
    "lst = []\n",
    "GSM_IDs = []\n",
    "index = 0\n",
    "for row in geo_df[\"Other matched data\"]:\n",
    "    \n",
    "    temp = []\n",
    "    state = False\n",
    "    \n",
    "    # Check for the term HiChIP in the title\n",
    "    for ele in row: \n",
    "        if (\"HiChIP\".casefold() in ele[\"Title\"].casefold()) or \\\n",
    "                (\"Hi-ChIP\".casefold() in ele[\"Title\"].casefold()):\n",
    "            state = True\n",
    "    \n",
    "    # keep samples with HiChIP and mark \"Yes\"\n",
    "    if state == True: \n",
    "        for ele in row:\n",
    "            GSM_IDs.append(ele[\"Accession\"])\n",
    "            if (\"HiChIP\".casefold() in ele[\"Title\"].casefold()) or \\\n",
    "                    (\"Hi-ChIP\".casefold() in ele[\"Title\"].casefold()):\n",
    "                temp.append(ele[\"Accession\"] + \": \" + ele[\"Title\"])\n",
    "        temp_str = \"\\n\".join(temp)\n",
    "        geo_df.at[index, \"Potential HiChIP\"] = \"Yes\"\n",
    "\n",
    "    # otherwise, keep all GSM samples and mark \"Maybe\"\n",
    "    else: \n",
    "        for ele in row:\n",
    "            GSM_IDs.append(ele[\"Accession\"])\n",
    "            temp.append(ele[\"Accession\"] + \": \" + ele[\"Title\"])\n",
    "        temp_str = \"\\n\".join(temp)\n",
    "        geo_df.at[index,\"Potential HiChIP\"]=\"Maybe\"\n",
    "    lst.append(temp_str)\n",
    "    index += 1\n",
    "    \n",
    "geo_df[\"Other matched data\"] = list(lst)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e88c1586",
   "metadata": {},
   "source": [
    "## Merge Rows with the Same Paper Title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4aa57284",
   "metadata": {},
   "outputs": [],
   "source": [
    "agg_functions = {'Paper Title':'first',\n",
    "                    'DOI':'max',\n",
    "                    'Journal':'first',\n",
    "                    'Authors':'first',\n",
    "                    'Year':'first',\n",
    "                    'GEO / Data link': lambda x: '\\n'.join(x),\n",
    "                    'Any other information':lambda x: '\\n'.join(x),\n",
    "                    'Organism':'first', 'Tissue/Cell Line':'first',\n",
    "                    'Potential HiChIP':'first',\n",
    "                    'Other matched data':lambda x: '\\n'.join(x),\n",
    "                    'Presenter':'first'}\n",
    "geo_df_grouped = geo_df.groupby(geo_df['Paper Title']).aggregate(agg_functions)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a13db4f0",
   "metadata": {},
   "source": [
    "## Fetch Journal Name and First Author of Each Paper and Add https to DOI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "55a1a9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### JR: Extremely slow, need to find a better way to do this\n",
    "\n",
    "# jounrnal_lst = []\n",
    "# authors_lst = []\n",
    "# fetch = PubMedFetcher()\n",
    "# for DOI in geo_df_grouped['DOI']:\n",
    "#     try:\n",
    "#         PMID = doi2pmid(DOI)\n",
    "#         article = fetch.article_by_pmid(PMID)\n",
    "#         jounrnal_lst.append(article.journal)\n",
    "#         authors_lst.append(article.authors[0].split()[0] + \" et al.\")\n",
    "#     except:\n",
    "#         jounrnal_lst.append(\"\")\n",
    "#         authors_lst.append(\"\")\n",
    "# geo_df_grouped['Journal'] = jounrnal_lst\n",
    "# geo_df_grouped['Authors'] = authors_lst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "9384d88e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # add https address to DOI (for easy access)\n",
    "# geo_df_grouped.loc[(geo_df_grouped['DOI'] != ''), 'DOI'] = 'https://doi.org/' + geo_df_grouped.loc[(geo_df_grouped['DOI'] != ''), 'DOI']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07a4b4c8",
   "metadata": {},
   "source": [
    "## Note the Date Added, Set Output Path, and Save the Final Output File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "beed3d3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make an output directory\n",
    "######## new for cell type\n",
    "outdir = ''\n",
    "# outdir = 'results/hichip_db/'\n",
    "# os.makedirs(outdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "be627b76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# determining the current year, month and day\n",
    "today = date.today()\n",
    "date_str = today.strftime(\"%Y_%m_%d\")\n",
    "\n",
    "# determining current hour and minute\n",
    "now = datetime.now()\n",
    "time_str = now.strftime(\"%H_%M\")\n",
    "\n",
    "# setting the output filename\n",
    "output = os.path.join(outdir, \"GEO_Query.{}_{}\".format(date_str, time_str))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "b1e2e626",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add a column that notes the added date\n",
    "geo_df_grouped['Date Added'] = date_str.replace('_', '-')\n",
    "reorder = ['Paper Title', 'Journal', 'Authors', 'Year', 'DOI', 'GEO / Data link', 'Any other information',\n",
    "           'Organism', 'Tissue/Cell Line', 'Potential HiChIP', 'Other matched data', 'Presenter', 'Date Added']\n",
    "geo_df_grouped = geo_df_grouped.loc[:, reorder]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "4d90d666",
   "metadata": {},
   "outputs": [],
   "source": [
    "geo_df_grouped.to_excel(output + \".xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea78b735",
   "metadata": {},
   "source": [
    "## Save the GSM IDs line by line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "3de96405",
   "metadata": {},
   "outputs": [],
   "source": [
    "GSM_IDs = list(set(GSM_IDs))\n",
    "with open(r'gsm_list.txt', 'w') as fp:\n",
    "    for gsm in GSM_IDs:\n",
    "        # write each item on a new line\n",
    "        fp.write(\"%s\\n\" % gsm)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8bbc7d4",
   "metadata": {},
   "source": [
    "## Query GEO BioSample using the GSM IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "49d34276",
   "metadata": {},
   "outputs": [],
   "source": [
    "GSM_IDs_from_file = []\n",
    "\n",
    "# open file and read the content in a list\n",
    "with open(r'gsm_list.txt', 'r') as fp:\n",
    "    for line in fp:\n",
    "        GSM_IDs_from_file.append(line)\n",
    "\n",
    "GSM_filters = ' [All Fields] OR '.join(GSM_IDs_from_file[4000:])\n",
    "\n",
    "\n",
    "search_result = Entrez.esearch(db=\"biosample\", retmax=10000, term=GSM_filters)\n",
    "result = Entrez.read(search_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8403720",
   "metadata": {},
   "source": [
    "### Filter query result by looping through XML"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7f1dce9",
   "metadata": {},
   "source": [
    "#### Dict of organs and their synynoms for classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "66cfe899",
   "metadata": {},
   "outputs": [],
   "source": [
    "json = OrderedDict()\n",
    "# order the classifcations from general to specific, so that as the code goes\n",
    "# down the ordered dict, general classifcations are replaced with specific ones\n",
    "json = {\n",
    "    \"biomaterial\": {\n",
    "        \"cell line\": [\"cellline\", \"cell line\", \"soxgfp\", \"cell line established from\", \"hESC\"], # pay attention to the last two\n",
    "        \"primary\": [\"primary\", \"primi\", \"attribute_name=\\\"patient\\\"\"]\n",
    "    },\n",
    "    \"disease\": {\n",
    "        \"HCC\": [\">hcc<\"],\n",
    "        \"GIST\": [\"gist\"],\n",
    "        \"Colorectal Adenocarcinoma\": [\"cw2\", \">rko\", \"sw403\", \"sw480\", \"sw620\", \"sw837\", \"sw948\", \"dld1\", \"hct15\", \"snu-c5\", \"hcc116\", \"ls174t\", \"colon cancer\", \"colorectal cancer\", \"coloO320-dm\"],\n",
    "        \"Leukemia\": [\"leukemia\"],\n",
    "        \"Acute myeloid leukemia\": [\"mll-af9\"],\n",
    "        \"Acute monocytic leukemia\": [\"thp-1\"],\n",
    "        \"Acute lymphoblastic leukaemia\": [\">sem\"],\n",
    "        \"Mantle cell lymphoma\": [\"mantle cell lymphoma\"],\n",
    "        \"ALL\": [\">all<\"],\n",
    "        \"Burkitt's lymphoma\": [\"ramos b\"],\n",
    "        \"Non-Hodgkin lymphoma\": [\"oci-ly7\", \"sudh4\"],\n",
    "        \"Prostate carcinoma\": [\"prostate cancer\", \"lncap\", \"vcap\"],\n",
    "        \"Normal\": [\">fhc\"],\n",
    "        \"Gastrointestinal stromal tumour\": [\"gist\"],\n",
    "        \"Mycosis fungoides\": [\"my-la\", \"myla\"],\n",
    "        \"Breast cancer\": [\"breast cancer\", \"t47d\"],\n",
    "        \"Invasive Breast Carcinoma\": [\"hcc1599\"],\n",
    "        \"Triple-negative breast cancer\": [\"triple-negative breast cancer\"],\n",
    "        \"ER-positive, PR-positive, HER2-negative breast cancer\": [\"mcf7\"],\n",
    "        \"ER-positive breast cancer\": [\"mda_mb134vi\"],\n",
    "        \"ER-positive, PR-negative and ERBB2-negative breast cancer\": [\"sum44pe\"],\n",
    "        \"Esophageal squamous cancer\": [\"esophageal squamous cancer\"],\n",
    "        \"Non-Small Cell Lung Cancer\": [\">hcc15<\"],\n",
    "        \"Lung squamous cancer\": [\"lung squamous cancer\"],\n",
    "        \"Small cell lung cancer\": [\"small cell lung cancer\"],\n",
    "        \"Head and neck squamous cancer\": [\"head and neck squamous cancer\"],\n",
    "        \"Human endometrial adenocarcinoma\": [\"human endometrial adenocarcinoma\"],\n",
    "        \"Ewing sarcoma\": [\"ewing sarcoma\"],\n",
    "        \"Multiple myeloma\": [\"multiple myeloma\"],\n",
    "        \"Neuroblastoma\": [\"neuroblastoma\", \"tr-14\", \"tr14\"],\n",
    "        \"Gastric carcinoma\": [\"snu16\", \"snu-16\"],\n",
    "        \"Rhabdoid tumor\": [\"g401\"],\n",
    "        \"Alveolar rhabdomyosarcoma\": [\">rh4<\", \">rh3<\", \">rh30<\"],\n",
    "        \"Embryonal rhabdomyosarcoma\": [\">sms\"],\n",
    "    },\n",
    "    # QUESTION: what do I do with ARMS, which has RH3, 30, 4, 41 metasized to different organs\n",
    "     # https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3713458/?\n",
    "    \"organ\": {\n",
    "        # ewing sarcoma are cells in the bones or connective tissue # plb might cause problem\n",
    "        \"embryo\": [\"embryo\", \"mesc\", \"hesc\", \"\\\">e14\", \"\\\">esc\"],\n",
    "        \"skeletal system\": [\"bone\", \"marrow\", \"ewing sarcoma\", \"t-all\", \"myeloma\", \"plb\", \"tc71\", \"\\\">rh30\", \"pre-b\"],\n",
    "        \"blood\": [\"vessel\", \"leukemia\", \"blood\", \"ramos b\", \"haematopoietic\", \"b-all\", \"\\\">sem\", \"jurkat\", \"cutll1\", \"gm12878\", \"oci-ly7\", \"thp-1\", \"hudep1\", \"hudep-1\", \"hudep2\", \"hudep-2\", \"hudep3\", \"hudep-3\"], # GSM6206927 bone or blood? # Hudep is from umbellical cord blood\n",
    "        \"bodily fluid\": [\"pleural\", \"effusion\", \"sudh4\"],\n",
    "        \"skin\": [\"skin\", \"epi\", \"squamous\", \"breast\", \"melanoma\", \"erythematosus\", \"sle\", \"keratinocyte\", \"melanocyte\", \"nasopharyngeal\"], # sle might cause problem\n",
    "        \"mammary gland\": [\"mammar\", \"breast\", \"mda-mb-231\"],\n",
    "        \"thyroid\": [\"thyroid gland\", \"thyroid\"],\n",
    "        \"thymus\": [\"thymus\", \"thymic\", \"treg\", \"t-reg\", \"my-la\", \"myla\"], # GSM5680696 thyroid or epithelium\n",
    "        \"lymph node\": [\"lymph node\", \"lymphoma\", \"\\\">rh3\\\"\"], # RH3 is derived from metastatic site, lymph node\n",
    "        \"tonsil\": [\"tonsil\"],\n",
    "        \"liver\": [\"liver\", \"hcc\"],\n",
    "        \"musculature of body\": [\"muscula\", \"soft\", \"connective\", \"rhabdomyosarcoma\", \"angiomatoid fibrous histiocytoma\", \"\\\">sms\", \"a673\"],\n",
    "        \"lung\": [\"lung\", \"\\\">rh4\"],\n",
    "        \"esophagus\": [\"esophagus\", \"esophageal\"], # GSM5680726 esophagus or epithelium\n",
    "        \"intestine\": [\"indestin\", \"fhc\", \"sw480\"],\n",
    "        \"kidney\": [\"kidney\", \"podocyte\", \"g401\"],\n",
    "        \"colon\": [\"colon\", \"colorectal\", \"hct116\", \"ls174t\", \">rko\", \"coloO320-dm\"],\n",
    "        \"spleen\": [\"spleen\"],\n",
    "        \"stomach\": [\"stomach\", \"gist\", \"snu16\", \"snu-16\"],\n",
    "        \"eye\": [\"retina\"],\n",
    "        \"placenta\": [\"placenta\"],\n",
    "        \"pancreas\": [\"pancrea\"],\n",
    "        \"bladder\": [\"bladder\"],\n",
    "        \"uturus\": [\"umbilical\", \"uterus\", \"endometrial\", \"cervic\", \"hela\"],\n",
    "        \"penis\": [\"penis\"],\n",
    "        \"testis\": [\"test\"],\n",
    "        \"ovary\": [\"ovar\"],\n",
    "        \"nerve\": [\"neuroblastoma\", \"nerve\", \"spinal\", \"cortex\", \"\\\">npc\", \"tr-14\", \"tr14\"],\n",
    "        \"brain\": [\"brain\", \"caudate\", \"chla-10\"],\n",
    "        \"prostate\": [\"prostate\"],\n",
    "        \"heart\": [\"heart\", \"cardi\", \"ventricle\", \"atrium\", \"aortic\"],\n",
    "        # dupliate the classifications except the synonyms are now cell line names or cell type names?\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bbda8c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 50 s per 100 GSMs\n",
    "# Initialize an empty dataframe with the desired column names\n",
    "cell_type_df = pd.DataFrame(columns=['name',\n",
    "                                     'organism',\n",
    "                                     'biomaterial',\n",
    "                                     'disease',\n",
    "                                     'organ',\n",
    "                                     'tissue',\n",
    "                                     'celltype',\n",
    "                                     'strain',\n",
    "                                     'sex',\n",
    "                                     'age',\n",
    "                                     'extdb_name',\n",
    "                                     'extdb_uuid',\n",
    "                                     'other'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "cf4dff12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100\n",
      "100 200\n",
      "200 300\n",
      "300 400\n",
      "400 500\n",
      "500 600\n",
      "600 700\n",
      "700 800\n",
      "800 900\n",
      "900 1000\n",
      "1000 1100\n",
      "1100 1200\n",
      "1200 1300\n",
      "1300 1400\n",
      "1400 1500\n",
      "1500 1600\n",
      "1600 1700\n",
      "1700 1800\n",
      "1800 1900\n",
      "1900 2000\n",
      "2000 2100\n",
      "2100 2200\n",
      "2200 2300\n",
      "2300 2400\n",
      "2400 2500\n",
      "2500 2600\n",
      "2600 2700\n",
      "2700 2800\n",
      "2800 2900\n",
      "2900 3000\n",
      "3000 3100\n",
      "3100 3200\n",
      "3200 3300\n",
      "3300 3400\n",
      "3400 3500\n",
      "3500 3600\n",
      "3600 3700\n",
      "3700 3800\n",
      "3800 3900\n",
      "3900 4000\n",
      "4000 4100\n",
      "4100 4200\n",
      "4200 4300\n",
      "4300 4400\n",
      "4400 4500\n",
      "4500 4600\n",
      "4600 4700\n",
      "4700 4800\n",
      "4800 4900\n",
      "4900 5000\n",
      "5000 5100\n",
      "5100 5200\n",
      "5200 5300\n"
     ]
    }
   ],
   "source": [
    "new_list = sorted(set(result[\"IdList\"]))\n",
    "\n",
    "# Define the base URL for the efetch command\n",
    "base_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi'\n",
    "\n",
    "root = None\n",
    "intervals = 100\n",
    "for start in range(0, len(new_list), intervals):\n",
    "\n",
    "    end = start + intervals\n",
    "    print(start, end)\n",
    "\n",
    "    # define the parameters for the efetch command\n",
    "    params = {\n",
    "        'db': 'biosample',\n",
    "        'id': ','.join(new_list[start:end]),\n",
    "        'retmode': 'xml',\n",
    "        'rettype': 'DocumentSummarySet'\n",
    "    }\n",
    "\n",
    "    # send the request to the E-utilities API and get the response\n",
    "    response = requests.get(base_url, params=params)\n",
    "\n",
    "    if root == None:\n",
    "        root = etree.fromstring(response.content)\n",
    "    else:\n",
    "        root2 = etree.fromstring(response.content)\n",
    "        root.extend(root2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "985ccf13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "finished GSM number 2600\n",
      "finished GSM number 2700\n",
      "finished GSM number 2800\n",
      "finished GSM number 2900\n",
      "finished GSM number 3000\n",
      "finished GSM number 3100\n",
      "finished GSM number 3200\n",
      "finished GSM number 3300\n",
      "finished GSM number 3400\n",
      "finished GSM number 3500\n",
      "finished GSM number 3600\n",
      "finished GSM number 3700\n",
      "finished GSM number 3800\n",
      "finished GSM number 3900\n",
      "finished GSM number 4000\n",
      "finished GSM number 4100\n",
      "finished GSM number 4200\n",
      "finished GSM number 4300\n",
      "finished GSM number 4400\n",
      "finished GSM number 4500\n",
      "finished GSM number 4600\n",
      "finished GSM number 4700\n",
      "finished GSM number 4800\n",
      "finished GSM number 4900\n",
      "finished GSM number 5000\n",
      "finished GSM number 5100\n",
      "finished GSM number 5200\n",
      "finished GSM number 5300\n",
      "finished GSM number 5400\n",
      "finished GSM number 5500\n",
      "finished GSM number 5600\n",
      "finished GSM number 5700\n",
      "finished GSM number 5800\n",
      "finished GSM number 5900\n",
      "finished GSM number 6000\n",
      "finished GSM number 6100\n",
      "finished GSM number 6200\n",
      "finished GSM number 6300\n",
      "finished GSM number 6400\n",
      "finished GSM number 6500\n",
      "finished GSM number 6600\n",
      "finished GSM number 6700\n",
      "finished GSM number 6800\n",
      "finished GSM number 6900\n",
      "finished GSM number 7000\n",
      "finished GSM number 7100\n",
      "finished GSM number 7200\n",
      "finished GSM number 7300\n",
      "finished GSM number 7400\n",
      "finished GSM number 7500\n",
      "finished GSM number 7600\n",
      "finished GSM number 7700\n"
     ]
    }
   ],
   "source": [
    "\n",
    "num_of_gsm_done = 0\n",
    "\n",
    "# Iterate through the list of GSM IDs\n",
    "for doc_sum in root.findall(\".//DocumentSummary\"):\n",
    "\n",
    "    data = doc_sum.find('.//SampleData')\n",
    "\n",
    "    gsm_data = {\n",
    "        \"name\":\"\",\n",
    "        \"organism\":\"\",\n",
    "        \"biomaterial\":\"\",\n",
    "        \"disease\": \"\",\n",
    "        \"organ\":\"\",\n",
    "        \"tissue\":\"\",\n",
    "        \"celltype\":\"\",\n",
    "        \"strain\":\"\",\n",
    "        \"sex\":\"\",\n",
    "        \"age\":\"\",\n",
    "        \"extdb_name\":\"\",\n",
    "        \"extdb_uuid\":\"\",\n",
    "        \"other\":\"\"\n",
    "        }\n",
    "    celltype_infos = []\n",
    "    \n",
    "    # Classify organ and biomaterial by checking synonyms in the SampleData XML text\n",
    "    # check GSM4956217 and GSM4447223\n",
    "    data_str = data.text\n",
    "    for dic in json:\n",
    "        for classification in json[dic]:\n",
    "            synonyms = json[dic][classification]\n",
    "            found = any(syn in data_str.lower() for syn in synonyms)\n",
    "            # edge cases\n",
    "            if \"\\\">RD\" in data_str:\n",
    "                gsm_data[\"organ\"] = \"musculature of body\"\n",
    "            if \"RH3\" in data_str:\n",
    "                gsm_data[\"organ\"] = \"exocrine gland\"\n",
    "            if classification == \"liver\": # sometimes hcc is actually breast, like hcc1599 and hcc19** cell lines\n",
    "                if \"breast\" in data_str.lower():\n",
    "                    gsm_data[\"organ\"] = \"breast\"\n",
    "            if found:\n",
    "                gsm_data[dic] = \"|\" + classification\n",
    "        gsm_data[dic] = gsm_data[dic].lstrip(\"|\")\n",
    "    \n",
    "    \n",
    "    for elem in data:\n",
    "        for subelem in elem:\n",
    "            if ('db\\': \\'GEO' in str(subelem.attrib)):\n",
    "                gsm_data[\"name\"] = subelem.text\n",
    "            if \"taxonomy_name\" in str(subelem.attrib):\n",
    "                gsm_data[\"organism\"] = subelem.attrib[\"taxonomy_name\"]\n",
    "            if ('disease' in str(subelem.attrib) or 'cancer type' in str(subelem.attrib)):\n",
    "                gsm_data[\"disease\"] = subelem.text # many organ info are from here\n",
    "            try:\n",
    "                if (gsm_data[\"disease\"] == \"\" and \\\n",
    "                    (\"oma\" in str(myroot.find(\"./Attributes/Attribute[@attribute_name='tissue']\").text).lower() or \\\n",
    "                    \"tumor\" in str(myroot.find(\"./Attributes/Attribute[@attribute_name='tissue']\").text).lower())\n",
    "                   ):\n",
    "                    gsm_data[\"disease\"] = myroot.find(\"./Attributes/Attribute[@attribute_name='tissue']\").text\n",
    "            except:\n",
    "                gsm_data[\"disease\"] = gsm_data[\"disease\"] + \"\"\n",
    "            if (gsm_data[\"disease\"] == \"\" and \\\n",
    "                (\"tumor\" in str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                \"tumour\" in str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                \"normal\" in str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower())\n",
    "               ):\n",
    "                    gsm_data[\"disease\"] = myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text\n",
    "            if ('organ' in str(subelem.attrib)): # no organ attributes at all\n",
    "                gsm_data[\"organ\"] = subelem.text\n",
    "            if ('tissue' in str(subelem.attrib) or \\\n",
    "                'brain region' in str(subelem.attrib)\n",
    "               ):\n",
    "                gsm_data[\"tissue\"] = subelem.text  # most organ info are from here # can have disease info\n",
    "            if ('cell_line' in str(subelem.attrib) or \\\n",
    "                'biological sample' in str(subelem.attrib) or \\\n",
    "                'neuronal subtype' in str(subelem.attrib) or \\\n",
    "                '\\'cell' in str(subelem.attrib)\n",
    "               ):\n",
    "                if ((not str(subelem.text).startswith(\"BRD\")) and (not str(subelem.text).startswith(\"MGH\"))): # these are gene names\n",
    "                    celltype_infos.append(str(subelem.attrib['attribute_name']) + \"-->\" + subelem.text)\n",
    "            if ('parental cell line' in str(subelem.attrib)):\n",
    "                celltype_infos.append(str(subelem.attrib['attribute_name']) + \"-->\" + subelem.text)\n",
    "                gsm_data[\"biomaterial\"] = \"cell line\"\n",
    "            if (\"mll-\" in str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                \"bl6\" in str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                \"hct\" in str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                \"vcap\" in str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                \"h9\" in str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                \"cell line\" in str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                (\"rim\" not in str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() and \\\n",
    "                 \"cyte\" in str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower())\n",
    "               ):\n",
    "                celltype_infos.append(\"{source_name}\" + \"-->\" + str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text))\n",
    "                gsm_data[\"biomaterial\"] = \"cell line\"\n",
    "            if ('cell_line' in str(subelem.attrib)):\n",
    "                if (subelem.text != \"--\"):\n",
    "                    gsm_data[\"biomaterial\"] = \"cell line\"\n",
    "            if ('strain' in str(subelem.attrib) or 'mouse model' in str(subelem.attrib)):\n",
    "                gsm_data[\"strain\"] = subelem.text\n",
    "            if ('p14' in str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower()):\n",
    "                gsm_data[\"strain\"] = str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text)\n",
    "            if ('gender' in str(subelem.attrib) or \\\n",
    "                '\\'sex\\'' in str(subelem.attrib)):\n",
    "                gsm_data[\"sex\"] = subelem.text\n",
    "            if (\"day\" in str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower()):\n",
    "                gsm_data[\"age\"] = str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text)\n",
    "            if ('\\'age\\'' in str(subelem.attrib) or \\\n",
    "                'age_day' in str(subelem.attrib)\n",
    "               ):\n",
    "                gsm_data[\"age\"] = subelem.text\n",
    "            if ('\\'time' in str(subelem.attrib) or \\\n",
    "                '\\'differentiation' in str(subelem.attrib) or \\\n",
    "                'differentiation\\'' in str(subelem.attrib) or \\\n",
    "                'developmental stage' in str(subelem.attrib) or \\\n",
    "                ('erythropoiesis' in str(subelem.attrib) and 'no' not in str(subelem.attrib))\n",
    "               ):\n",
    "                if gsm_data[\"age\"] == \"\":\n",
    "                    gsm_data[\"age\"] = subelem.text\n",
    "            \n",
    "    gsm_data[\"celltype\"] = '|'.join(set(celltype_infos))\n",
    "    gsm_data[\"other\"] = \"{source_name}\" + \"-->\" + str(myroot.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text)\n",
    "    \n",
    "    cell_type_df = cell_type_df.append(gsm_data, ignore_index=True)\n",
    "    \n",
    "    # indicate how many GSM are processed and save the table as the loop runs\n",
    "    num_of_gsm_done += 1\n",
    "    if (num_of_gsm_done % 100 == 0):\n",
    "            print(\"finished GSM number\", num_of_gsm_done)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71d8fecb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "916b09ad",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "972816a3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "203d163c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63e617d9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c298725d",
   "metadata": {},
   "source": [
    "### Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d24f12b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output file:  GEO_Query_cell_type.2022_12_23_18_16\n"
     ]
    }
   ],
   "source": [
    "# setting the output filename\n",
    "today = date.today()\n",
    "date_str = today.strftime(\"%Y_%m_%d\")\n",
    "now = datetime.now()\n",
    "time_str = now.strftime(\"%H_%M\")\n",
    "output = \"GEO_Query_cell_type.{}_{}\".format(date_str, time_str)\n",
    "print(\"output file: \", output)\n",
    "cell_type_df.to_excel(output + \".xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5a70117",
   "metadata": {},
   "source": [
    "Sample ID: This is the unique identifier for each sample in the database. You can search for samples by their ID using this field.\n",
    "\n",
    "Organism: This field allows you to search for samples from a specific organism, such as human, mouse, or yeast.\n",
    "\n",
    "Tissue: This field allows you to search for samples based on their tissue type, such as liver, brain, or heart.\n",
    "\n",
    "Disease state: This field allows you to search for samples based on their disease state, such as cancer, diabetes, or autoimmune disease.\n",
    "\n",
    "Cell type: This field allows you to search for samples based on their cell type, such as fibroblasts, macrophages, or stem cells.\n",
    "\n",
    "Developmental stage: This field allows you to search for samples based on their developmental stage, such as embryonic, fetal, or adult.\n",
    "\n",
    "Experimental design: This field allows you to search for samples based on the type of experiment they were used in, such as microarray, RNA-seq, or ChIP-seq."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  },
  "vscode": {
   "interpreter": {
    "hash": "036b15a84498062e1e205bf5f2012dec27d9ce09ecb8b5e207c5080560abe931"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
