{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ee429c1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-02-24 16:00:17 herman-login3.hpc.lji.org metapub.config[12320] WARNING NCBI_API_KEY was not set.\n"
     ]
    }
   ],
   "source": [
    "import os \n",
    "import re\n",
    "import requests\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "import ssl\n",
    "ssl._create_default_https_context = ssl._create_unverified_context\n",
    "\n",
    "from datetime import date\n",
    "from datetime import datetime\n",
    "\n",
    "from metapub.convert import doi2pmid\n",
    "from metapub import PubMedFetcher\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from lxml import etree\n",
    "from Bio import Entrez\n",
    "\n",
    "# make an output directory\n",
    "######## new for cell type\n",
    "outdir = '../../results/geo/\n",
    "# outdir = 'results/hichip_db/'\n",
    "# os.makedirs(outdir, exist_ok=True)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1ac51ef2",
   "metadata": {},
   "source": [
    "## Query the GEO database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fbd0ff62",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use one of the following search queries/filters\n",
    "# filters = \"HiChIP[All Fields] AND \\\"gse\\\"[Filter]\" # the whole notebook will take around 7 min to run\n",
    "# filters = \"HiChIP[All Fields] AND (\\\"gse\\\"[Filter] AND (\\\"Homo sapiens\\\"[Organism] OR \\\"Mus musculus\\\"[Organism]))\n",
    "# filters = \"HiChIP[All Fields] AND (\\\"Homo sapiens\\\"[Organism] AND \\\"gse\\\"[Filter])\"\n",
    "# filters = \"HiChIP[All Fields] AND (\\\"Homo sapiens\\\"[Organism] AND \\\"published last year\\\"[Filter]) AND \\\"gse\\\"[Filter]\" # 2 min\n",
    "# filters = \"HiChIP[All Fields] AND (\\\"Mus musculus\\\"[Organism] AND \\\"gse\\\"[Filter])\"\n",
    "\n",
    "# set a dummy email \n",
    "Entrez.email = \"zjiang@lji.org\"\n",
    "\n",
    "# query the GEO database for HiChIP samples derived from human or mouse\n",
    "# note: gds is an abbreviation for Gene Expression Omnibus (GEO) database\n",
    "filters = \"HiChIP[All Fields] AND (\\\"gse\\\"[Filter] AND (\\\"Homo sapiens\\\"[Organism] OR \\\"Mus musculus\\\"[Organism]))\"\n",
    "geo_query = Entrez.esearch(db=\"gds\", retmax=10000, term=filters)\n",
    "geo_results = Entrez.read(geo_query)\n",
    "\n",
    "# get summary of these entries\n",
    "geo_ids = ','.join(geo_results['IdList'])\n",
    "geo_summary = Entrez.esummary(db=\"gds\", id=geo_ids, retmode=\"xml\") \n",
    "geo_entries = list(Entrez.parse(geo_summary))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "bb4d1e58",
   "metadata": {},
   "source": [
    "## Obtaining a Map Between PMID and DOI"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "cd1684e3",
   "metadata": {},
   "source": [
    "#### Extract PMIDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eb8369c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['27643841',\n",
       " '28945252',\n",
       " '29224777',\n",
       " '29641996',\n",
       " '29706538',\n",
       " '29731168',\n",
       " '30046115',\n",
       " '30122536',\n",
       " '30397335',\n",
       " '30580963']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pubmed_ids = set()\n",
    "for entry in geo_entries:\n",
    "    for value in entry['PubMedIds']:\n",
    "        pubmed_ids.add(str(int(value)))\n",
    "pubmed_ids = sorted(pubmed_ids)\n",
    "pubmed_ids[0:10]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "975347e7",
   "metadata": {},
   "source": [
    "#### Convert from PMID to DOI using the eutils API"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "119333b2",
   "metadata": {},
   "source": [
    "The API was used directly since it was faster than using the metapub package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "63673aca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the base URL for the efetch command\n",
    "efetch_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi'\n",
    "\n",
    "# define the parameters for the efetch command\n",
    "params = {\n",
    "    'db': 'pubmed',\n",
    "    'id': ','.join(pubmed_ids),\n",
    "    'retmode': 'xml',\n",
    "    'rettype': 'docsum'\n",
    "}\n",
    "\n",
    "# send the request to the E-utilities API and get the response\n",
    "ds_response = requests.get(efetch_url, params=params)\n",
    "\n",
    "# parse the XML response and extract the DOIs\n",
    "ds_root = etree.fromstring(ds_response.content)\n",
    "doc_sums = ds_root.findall('DocSum')\n",
    "pubmed_data = []\n",
    "for doc_sum in doc_sums:\n",
    "\n",
    "    pmid = doc_sum.find('Id').text\n",
    "    dois = [x.text for x in doc_sum.findall('Item[@Name=\"DOI\"]')]\n",
    "    dois = ','.join(dois)\n",
    "    pubmed_data.append([pmid, dois])\n",
    "\n",
    "pubmed_data = pd.DataFrame(pubmed_data, columns=['PMID', 'DOI'])\n",
    "pubmed_data.set_index('PMID', inplace=True)\n",
    "pubmed_data = pubmed_data.squeeze()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "5f1bbb57",
   "metadata": {},
   "source": [
    "## Convert GEO entries into a dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4c5feeac",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "geo_data = []\n",
    "for i, entry in enumerate(geo_entries):\n",
    "\n",
    "    ##### Cleaning up the title #####\n",
    "    paper_title = entry['title']\n",
    "    new_paper_title = ''\n",
    "    last_word = re.split('\\s|[.]', paper_title)[-1]\n",
    "    last_two_words = re.split('\\s|[.]', paper_title)[-2:]\n",
    "    \n",
    "    # remove \".[HiChIP]\", \". [HiChIP]\", or \"(HiChIP)\" in paper titles using steps below\n",
    "    if ('[' in last_word and ']' in last_word) or \\\n",
    "            ('(' in last_word and ')' in last_word):\n",
    "        new_paper_title = re.split('\\s|[.]', paper_title)[0:-1]\n",
    "        new_paper_title = ' '.join(new_paper_title)\n",
    "        new_paper_title = new_paper_title.strip()\n",
    "\n",
    "        # remove the last period in paper title\n",
    "        if new_paper_title.endswith('.'):\n",
    "            new_paper_title = new_paper_title[:-1]\n",
    "            \n",
    "    # remove \".[Hi ChIP]\", \". [Hi ChIP]\", or \"[Bead Array]\" in paper titles using steps below\n",
    "    elif ('[' in last_two_words[0] and ']' in last_two_words[1]) or \\\n",
    "                ('(' in last_two_words[0] and ')' in last_two_words[1]):\n",
    "        new_paper_title = re.split('\\s|[.]', paper_title)[0:-2]\n",
    "        new_paper_title = ' '.join(new_paper_title)\n",
    "        new_paper_title = new_paper_title.strip()\n",
    "\n",
    "        # remove the last period in paper title\n",
    "        if new_paper_title.endswith('.'):\n",
    "            new_paper_title = new_paper_title[:-1]\n",
    "            \n",
    "    # remove the period\n",
    "    elif paper_title.endswith('.'):\n",
    "        new_paper_title = paper_title[:-1] \n",
    "        \n",
    "    else:\n",
    "        new_paper_title = paper_title\n",
    "\n",
    "    new_paper_title = new_paper_title.replace(\"\\xa0\", \" \") # remove no-break space\n",
    "    entry['title'] = new_paper_title\n",
    "\n",
    "    ##### add the doi to the entry #####\n",
    "    curr_dois = []\n",
    "    for value in entry['PubMedIds']:\n",
    "        pubmed_id = str(int(value))\n",
    "        curr_doi = pubmed_data[pubmed_id]\n",
    "        curr_dois.append(curr_doi)\n",
    "    entry['DOI'] = ','.join(curr_dois)\n",
    "\n",
    "    ##### add the adjusted entry to the data #####\n",
    "    geo_data.append(entry)\n",
    "    \n",
    "geo_df = pd.DataFrame(geo_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b7fbf94b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Item</th>\n",
       "      <th>Id</th>\n",
       "      <th>Accession</th>\n",
       "      <th>GDS</th>\n",
       "      <th>title</th>\n",
       "      <th>summary</th>\n",
       "      <th>GPL</th>\n",
       "      <th>GSE</th>\n",
       "      <th>taxon</th>\n",
       "      <th>entryType</th>\n",
       "      <th>...</th>\n",
       "      <th>n_samples</th>\n",
       "      <th>SeriesTitle</th>\n",
       "      <th>PlatformTitle</th>\n",
       "      <th>PlatformTaxa</th>\n",
       "      <th>SamplesTaxa</th>\n",
       "      <th>PubMedIds</th>\n",
       "      <th>Projects</th>\n",
       "      <th>FTPLink</th>\n",
       "      <th>GEO2R</th>\n",
       "      <th>DOI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[]</td>\n",
       "      <td>200223259</td>\n",
       "      <td>GSE223259</td>\n",
       "      <td></td>\n",
       "      <td>Chromatin accessibilities in C9HRE ALS/FTD pat...</td>\n",
       "      <td>This SuperSeries is composed of the SubSeries ...</td>\n",
       "      <td>16791</td>\n",
       "      <td>223259</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>24</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE223nn...</td>\n",
       "      <td>yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[]</td>\n",
       "      <td>200223258</td>\n",
       "      <td>GSE223258</td>\n",
       "      <td></td>\n",
       "      <td>To profile the chromatin structural changes re...</td>\n",
       "      <td>DAXX is an epigenetic regulator and chromatin ...</td>\n",
       "      <td>16791</td>\n",
       "      <td>223258</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE223nn...</td>\n",
       "      <td>yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[]</td>\n",
       "      <td>200178673</td>\n",
       "      <td>GSE178673</td>\n",
       "      <td></td>\n",
       "      <td>Chromatin landscape in perinatal cardiomyocytes</td>\n",
       "      <td>In order to understand the transcriptional reg...</td>\n",
       "      <td>21273;24688</td>\n",
       "      <td>178673</td>\n",
       "      <td>Mus musculus; Rattus norvegicus</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>28</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[IntegerElement(36653336, attributes={})]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE178nn...</td>\n",
       "      <td>yes</td>\n",
       "      <td>10.1038/s41420-023-01322-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[]</td>\n",
       "      <td>200215020</td>\n",
       "      <td>GSE215020</td>\n",
       "      <td></td>\n",
       "      <td>Generate 3D genome database for atrial and ven...</td>\n",
       "      <td>Measurement the contribution of 3D genome stru...</td>\n",
       "      <td>24247</td>\n",
       "      <td>215020</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE215nn...</td>\n",
       "      <td>yes</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[]</td>\n",
       "      <td>200215065</td>\n",
       "      <td>GSE215065</td>\n",
       "      <td></td>\n",
       "      <td>Regulation of mouse chamber selective enhancers</td>\n",
       "      <td>This SuperSeries is composed of the SubSeries ...</td>\n",
       "      <td>19057;26526;24247</td>\n",
       "      <td>215065</td>\n",
       "      <td>synthetic construct; Mus musculus</td>\n",
       "      <td>GSE</td>\n",
       "      <td>...</td>\n",
       "      <td>82</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[IntegerElement(36705030, attributes={})]</td>\n",
       "      <td>[]</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE215nn...</td>\n",
       "      <td>yes</td>\n",
       "      <td>10.1161/CIRCULATIONAHA.122.061955</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Item         Id  Accession GDS  \\\n",
       "0   []  200223259  GSE223259       \n",
       "1   []  200223258  GSE223258       \n",
       "2   []  200178673  GSE178673       \n",
       "3   []  200215020  GSE215020       \n",
       "4   []  200215065  GSE215065       \n",
       "\n",
       "                                               title  \\\n",
       "0  Chromatin accessibilities in C9HRE ALS/FTD pat...   \n",
       "1  To profile the chromatin structural changes re...   \n",
       "2    Chromatin landscape in perinatal cardiomyocytes   \n",
       "3  Generate 3D genome database for atrial and ven...   \n",
       "4    Regulation of mouse chamber selective enhancers   \n",
       "\n",
       "                                             summary                GPL  \\\n",
       "0  This SuperSeries is composed of the SubSeries ...              16791   \n",
       "1  DAXX is an epigenetic regulator and chromatin ...              16791   \n",
       "2  In order to understand the transcriptional reg...        21273;24688   \n",
       "3  Measurement the contribution of 3D genome stru...              24247   \n",
       "4  This SuperSeries is composed of the SubSeries ...  19057;26526;24247   \n",
       "\n",
       "      GSE                              taxon entryType  ... n_samples  \\\n",
       "0  223259                       Homo sapiens       GSE  ...        24   \n",
       "1  223258                       Homo sapiens       GSE  ...         3   \n",
       "2  178673    Mus musculus; Rattus norvegicus       GSE  ...        28   \n",
       "3  215020                       Mus musculus       GSE  ...         6   \n",
       "4  215065  synthetic construct; Mus musculus       GSE  ...        82   \n",
       "\n",
       "  SeriesTitle PlatformTitle PlatformTaxa SamplesTaxa  \\\n",
       "0                                                      \n",
       "1                                                      \n",
       "2                                                      \n",
       "3                                                      \n",
       "4                                                      \n",
       "\n",
       "                                   PubMedIds Projects  \\\n",
       "0                                         []       []   \n",
       "1                                         []       []   \n",
       "2  [IntegerElement(36653336, attributes={})]       []   \n",
       "3                                         []       []   \n",
       "4  [IntegerElement(36705030, attributes={})]       []   \n",
       "\n",
       "                                             FTPLink GEO2R  \\\n",
       "0  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE223nn...   yes   \n",
       "1  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE223nn...   yes   \n",
       "2  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE178nn...   yes   \n",
       "3  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE215nn...   yes   \n",
       "4  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE215nn...   yes   \n",
       "\n",
       "                                 DOI  \n",
       "0                                     \n",
       "1                                     \n",
       "2         10.1038/s41420-023-01322-3  \n",
       "3                                     \n",
       "4  10.1161/CIRCULATIONAHA.122.061955  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geo_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c463820d",
   "metadata": {},
   "source": [
    "## Reformat the Columns for Google Sheet Compatibility"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e5bc08ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop columns\n",
    "drop_cols = ['Item', 'Id', 'GDS', 'GPL', 'GSE', 'entryType', 'ptechType',\n",
    "             'valType', 'SSInfo', 'subsetInfo', 'suppFile', 'Relations', 'ExtRelations',\n",
    "             'n_samples', 'SeriesTitle', 'PlatformTitle', 'PlatformTaxa', 'SamplesTaxa',\n",
    "             'Projects', 'FTPLink', 'GEO2R']\n",
    "geo_df.drop(drop_cols, inplace=True, axis=1)\n",
    "\n",
    "# rename columns\n",
    "rename_cols = {\"Accession\":\"GEO / Data link\",\n",
    "               \"title\": \"Paper Title\",\n",
    "               \"taxon\": \"Organism\",\n",
    "               \"gdsType\": \"Any other information\",\n",
    "               \"PDAT\": \"Year\",\n",
    "               \"Samples\": \"Other matched data\"}\n",
    "geo_df.rename(columns=rename_cols, inplace=True)\n",
    "\n",
    "# extract just the year\n",
    "geo_df[\"Year\"] = geo_df[\"Year\"].str[:4]\n",
    "\n",
    "# add index for merging\n",
    "geo_df['index'] = np.arange(len(geo_df))\n",
    "\n",
    "# fill na with empty values\n",
    "geo_df.fillna(\"\", inplace=True)\n",
    "\n",
    "# adding missing columns with empty values\n",
    "geo_df[\"Journal\"] = \"\"\n",
    "geo_df[\"Authors\"] = \"\"\n",
    "geo_df[\"Tissue/Cell Line\"] = \"\"\n",
    "geo_df[\"Presenter\"] = \"\"\n",
    "geo_df[\"Potential HiChIP\"] = \"\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "19c61b12",
   "metadata": {},
   "source": [
    "## Assigning Labels to Indicate a Potential HiChIP Sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ecfe777c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out samples with no HiChIP mention and assigned Yes or Maybe to all others \n",
    "lst = []\n",
    "GSM_IDs = []\n",
    "index = 0\n",
    "for row in geo_df[\"Other matched data\"]:\n",
    "    \n",
    "    temp = []\n",
    "    state = False\n",
    "    \n",
    "    # Check for the term HiChIP in the title\n",
    "    for ele in row: \n",
    "        if (\"HiChIP\".casefold() in ele[\"Title\"].casefold()) or \\\n",
    "                (\"Hi-ChIP\".casefold() in ele[\"Title\"].casefold()):\n",
    "            state = True\n",
    "    \n",
    "    # keep samples with HiChIP and mark \"Yes\"\n",
    "    if state == True: \n",
    "        for ele in row:\n",
    "            GSM_IDs.append(ele[\"Accession\"])\n",
    "            if (\"HiChIP\".casefold() in ele[\"Title\"].casefold()) or \\\n",
    "                    (\"Hi-ChIP\".casefold() in ele[\"Title\"].casefold()):\n",
    "                temp.append(ele[\"Accession\"] + \": \" + ele[\"Title\"])\n",
    "        temp_str = \"\\n\".join(temp)\n",
    "        geo_df.at[index, \"Potential HiChIP\"] = \"Yes\"\n",
    "\n",
    "    # otherwise, keep all GSM samples and mark \"Maybe\"\n",
    "    else: \n",
    "        for ele in row:\n",
    "            GSM_IDs.append(ele[\"Accession\"])\n",
    "            temp.append(ele[\"Accession\"] + \": \" + ele[\"Title\"])\n",
    "        temp_str = \"\\n\".join(temp)\n",
    "        geo_df.at[index,\"Potential HiChIP\"]=\"Maybe\"\n",
    "    lst.append(temp_str)\n",
    "    index += 1\n",
    "    \n",
    "geo_df[\"Other matched data\"] = list(lst)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e88c1586",
   "metadata": {},
   "source": [
    "## Merge Rows with the Same Paper Title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4aa57284",
   "metadata": {},
   "outputs": [],
   "source": [
    "agg_functions = {'Paper Title':'first',\n",
    "                    'DOI':'max',\n",
    "                    'Journal':'first',\n",
    "                    'Authors':'first',\n",
    "                    'Year':'first',\n",
    "                    'GEO / Data link': lambda x: '\\n'.join(x),\n",
    "                    'Any other information':lambda x: '\\n'.join(x),\n",
    "                    'Organism':'first', 'Tissue/Cell Line':'first',\n",
    "                    'Potential HiChIP':'first',\n",
    "                    'Other matched data':lambda x: '\\n'.join(x),\n",
    "                    'Presenter':'first'}\n",
    "geo_df_grouped = geo_df.groupby(geo_df['Paper Title']).aggregate(agg_functions)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a13db4f0",
   "metadata": {},
   "source": [
    "## Fetch Additional Paper Metadata"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "6bf63918",
   "metadata": {},
   "source": [
    "Adding journal name, first author and add https to DOIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "55a1a9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### JR: Extremely slow, need to find a better way to do this\n",
    "\n",
    "# jounrnal_lst = []\n",
    "# authors_lst = []\n",
    "# fetch = PubMedFetcher()\n",
    "# for DOI in geo_df_grouped['DOI']:\n",
    "#     try:\n",
    "#         PMID = doi2pmid(DOI)\n",
    "#         article = fetch.article_by_pmid(PMID)\n",
    "#         jounrnal_lst.append(article.journal)\n",
    "#         authors_lst.append(article.authors[0].split()[0] + \" et al.\")\n",
    "#     except:\n",
    "#         jounrnal_lst.append(\"\")\n",
    "#         authors_lst.append(\"\")\n",
    "# geo_df_grouped['Journal'] = jounrnal_lst\n",
    "# geo_df_grouped['Authors'] = authors_lst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "9384d88e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # add https address to DOI (for easy access)\n",
    "# geo_df_grouped.loc[(geo_df_grouped['DOI'] != ''), 'DOI'] = 'https://doi.org/' + geo_df_grouped.loc[(geo_df_grouped['DOI'] != ''), 'DOI']"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "07a4b4c8",
   "metadata": {},
   "source": [
    "## Save the Final Output File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "be627b76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add a column that notes the added date\n",
    "geo_df_grouped['Date Added'] = date_str.replace('_', '-')\n",
    "reorder = ['Paper Title', 'Journal', 'Authors', 'Year', 'DOI',\n",
    "           'GEO / Data link', 'Any other information',\n",
    "           'Organism', 'Tissue/Cell Line', 'Potential HiChIP',\n",
    "           'Other matched data', 'Presenter', 'Date Added']\n",
    "geo_df_grouped = geo_df_grouped.loc[:, reorder]\n",
    "\n",
    "# determining the current year, month and day\n",
    "today = date.today()\n",
    "date_str = today.strftime(\"%Y_%m_%d\")\n",
    "\n",
    "# # determining current hour and minute\n",
    "# now = datetime.now()\n",
    "# time_str = now.strftime(\"%H_%M\")\n",
    "\n",
    "# setting the output filename\n",
    "#output = os.path.join(outdir, \"geo.query.{}_{}.xlsx\".format(date_str, time_str))\n",
    "output = os.path.join(outdir, \"geo.query.{}.xlsx\".format(date_str))\n",
    "geo_df_grouped.to_excel(output, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  },
  "vscode": {
   "interpreter": {
    "hash": "036b15a84498062e1e205bf5f2012dec27d9ce09ecb8b5e207c5080560abe931"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
