{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ee429c1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "import ssl\n",
    "ssl._create_default_https_context = ssl._create_unverified_context\n",
    "import requests\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from lxml import etree\n",
    "from Bio import Entrez\n",
    "\n",
    "import json\n",
    "\n",
    "# set a dummy email \n",
    "Entrez.email = \"zjiang@lji.org\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8bbc7d4",
   "metadata": {},
   "source": [
    "## Query GEO BioSample using the GSM IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "49d34276",
   "metadata": {},
   "outputs": [],
   "source": [
    "# open file and read the content in a list\n",
    "gsm_ids = []\n",
    "with open('../geo_queries/gsm_list.txt', 'r') as fp:\n",
    "    gsm_ids = [x.strip() for x in fp.readlines()]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "24a6ff74",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsm_filters = '[All Fields] OR '.join(gsm_ids)\n",
    "gsm_query = Entrez.esearch(db=\"biosample\", retmax=10000, term=gsm_filters)\n",
    "gsm_result = Entrez.read(gsm_query)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8403720",
   "metadata": {},
   "source": [
    "### Filter query result by looping through XML"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c7f1dce9",
   "metadata": {},
   "source": [
    "#### Define a dictionary of organs and their synynoms for classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "66cfe899",
   "metadata": {},
   "outputs": [],
   "source": [
    "# order the classifcations from general to specific, so that as the code goes\n",
    "# down the ordered dict, general classifcations are replaced with specific ones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e9304ba8",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('metadata_dictionary.json', 'r') as fp:\n",
    "    class_dict = json.load(fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1bbda8c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize an empty dataframe with the desired column names\n",
    "cell_type_df = pd.DataFrame(columns=['name', 'organism', 'biomaterial', 'disease',\n",
    "                                     'organ', 'tissue', 'celltype', 'strain',\n",
    "                                     'sex', 'age', 'extdb_name', 'extdb_uuid', 'other'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "cf4dff12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 100\n",
      "100 200\n",
      "200 300\n",
      "300 400\n",
      "400 500\n",
      "500 600\n",
      "600 700\n",
      "700 800\n",
      "800 900\n",
      "900 1000\n",
      "1000 1100\n",
      "1100 1200\n"
     ]
    }
   ],
   "source": [
    "gsm_id_list = sorted(set(gsm_result[\"IdList\"]))\n",
    "\n",
    "# Define the base URL for the efetch command\n",
    "base_url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi'\n",
    "\n",
    "root = None\n",
    "intervals = 100\n",
    "for start in range(0, len(gsm_id_list), intervals):\n",
    "\n",
    "    end = start + intervals\n",
    "    print(start, end)\n",
    "\n",
    "    # define the parameters for the efetch command\n",
    "    params = {\n",
    "        'db': 'biosample',\n",
    "        'id': ','.join(gsm_id_list[start:end]),\n",
    "        'retmode': 'xml',\n",
    "        'rettype': 'DocumentSummarySet'\n",
    "    }\n",
    "\n",
    "    # send the request to the E-utilities API and get the response\n",
    "    response = requests.get(base_url, params=params)\n",
    "\n",
    "    if root == None:\n",
    "        root = etree.fromstring(response.content)\n",
    "    else:\n",
    "        root2 = etree.fromstring(response.content)\n",
    "        root.extend(root2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a563ce07",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbb103ec",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "293aad6b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88eabd9f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "985ccf13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "finished GSM number 100\n",
      "finished GSM number 200\n",
      "finished GSM number 300\n",
      "finished GSM number 400\n",
      "finished GSM number 500\n",
      "finished GSM number 600\n",
      "finished GSM number 700\n",
      "finished GSM number 800\n",
      "finished GSM number 900\n",
      "finished GSM number 1000\n",
      "finished GSM number 1100\n",
      "finished GSM number 1200\n",
      "finished GSM number 1300\n",
      "finished GSM number 1400\n",
      "finished GSM number 1500\n",
      "finished GSM number 1600\n",
      "finished GSM number 1700\n",
      "finished GSM number 1800\n",
      "finished GSM number 1900\n",
      "finished GSM number 2000\n",
      "finished GSM number 2100\n",
      "finished GSM number 2200\n",
      "finished GSM number 2300\n",
      "finished GSM number 2400\n",
      "finished GSM number 2500\n",
      "finished GSM number 2600\n",
      "finished GSM number 2700\n",
      "finished GSM number 2800\n",
      "finished GSM number 2900\n",
      "finished GSM number 3000\n",
      "finished GSM number 3100\n",
      "finished GSM number 3200\n",
      "finished GSM number 3300\n",
      "finished GSM number 3400\n",
      "finished GSM number 3500\n",
      "finished GSM number 3600\n",
      "finished GSM number 3700\n",
      "finished GSM number 3800\n",
      "finished GSM number 3900\n",
      "finished GSM number 4000\n",
      "finished GSM number 4100\n",
      "finished GSM number 4200\n",
      "finished GSM number 4300\n",
      "finished GSM number 4400\n",
      "finished GSM number 4500\n",
      "finished GSM number 4600\n",
      "finished GSM number 4700\n",
      "finished GSM number 4800\n",
      "finished GSM number 4900\n",
      "finished GSM number 5000\n",
      "finished GSM number 5100\n",
      "finished GSM number 5200\n",
      "finished GSM number 5300\n",
      "finished GSM number 5400\n",
      "finished GSM number 5500\n",
      "finished GSM number 5600\n",
      "finished GSM number 5700\n",
      "finished GSM number 5800\n",
      "finished GSM number 5900\n",
      "finished GSM number 6000\n",
      "finished GSM number 6100\n",
      "finished GSM number 6200\n",
      "finished GSM number 6300\n",
      "finished GSM number 6400\n",
      "finished GSM number 6500\n",
      "finished GSM number 6600\n",
      "finished GSM number 6700\n",
      "finished GSM number 6800\n",
      "finished GSM number 6900\n",
      "finished GSM number 7000\n",
      "finished GSM number 7100\n",
      "finished GSM number 7200\n",
      "finished GSM number 7300\n",
      "finished GSM number 7400\n",
      "finished GSM number 7500\n",
      "finished GSM number 7600\n",
      "finished GSM number 7700\n",
      "finished GSM number 7800\n",
      "finished GSM number 7900\n",
      "finished GSM number 8000\n",
      "finished GSM number 8100\n",
      "finished GSM number 8200\n",
      "finished GSM number 8300\n",
      "finished GSM number 8400\n",
      "finished GSM number 8500\n",
      "finished GSM number 8600\n",
      "finished GSM number 8700\n",
      "finished GSM number 8800\n",
      "finished GSM number 8900\n",
      "finished GSM number 9000\n",
      "finished GSM number 9100\n"
     ]
    }
   ],
   "source": [
    "num_of_gsm_done = 0\n",
    "\n",
    "# Iterate through the list of GSM IDs\n",
    "for doc_sum in root.findall(\".//DocumentSummary\"):\n",
    "\n",
    "    sample_data = doc_sum.find('.//SampleData')\n",
    "\n",
    "    gsm_data = {\n",
    "        'name': '',\n",
    "        'organism': '',\n",
    "        'biomaterial': '',\n",
    "        'disease': '',\n",
    "        'organ': '',\n",
    "        'tissue': '',\n",
    "        'celltype': '',\n",
    "        'strain': '',\n",
    "        'sex': '',\n",
    "        'age': '',\n",
    "        'extdb_name': '',\n",
    "        'extdb_uuid': '',\n",
    "        'other': ''}\n",
    "    celltype_infos = []\n",
    "    \n",
    "    # Classify organ and biomaterial by checking synonyms in the SampleData XML text\n",
    "    data_str = sample_data.text\n",
    "    for key in class_dict:\n",
    "        for classification in class_dict[key]:\n",
    "            synonyms = class_dict[key][classification]\n",
    "            found = any(syn in data_str.lower() for syn in synonyms)\n",
    "\n",
    "            # edge cases\n",
    "            if \"\\\">RD\" in data_str:\n",
    "                gsm_data[\"organ\"] = \"musculature of body\"\n",
    "\n",
    "            if \"RH3\" in data_str:\n",
    "                gsm_data[\"organ\"] = \"exocrine gland\"\n",
    "\n",
    "            if classification == \"liver\": # sometimes hcc is actually breast, like hcc1599 and hcc19** cell lines\n",
    "                if \"breast\" in data_str.lower():\n",
    "                    gsm_data[\"organ\"] = \"breast\"\n",
    "\n",
    "            if found:\n",
    "                gsm_data[key] = \"|\" + classification\n",
    "                \n",
    "        gsm_data[key] = gsm_data[key].lstrip(\"|\")\n",
    "    \n",
    "    for elem in sample_data:\n",
    "        for subelem in elem:\n",
    "\n",
    "            if ('db\\': \\'GEO' in str(subelem.attrib)):\n",
    "                gsm_data[\"name\"] = subelem.text\n",
    "\n",
    "            if \"taxonomy_name\" in str(subelem.attrib):\n",
    "                gsm_data[\"organism\"] = subelem.attrib[\"taxonomy_name\"]\n",
    "\n",
    "            if ('disease' in str(subelem.attrib) or 'cancer type' in str(subelem.attrib)):\n",
    "                gsm_data[\"disease\"] = subelem.text # many organ info are from here\n",
    "\n",
    "            try:\n",
    "                if (gsm_data[\"disease\"] == \"\" and \\\n",
    "                    (\"oma\" in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='tissue']\").text).lower() or \\\n",
    "                    \"tumor\" in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='tissue']\").text).lower())\n",
    "                   ):\n",
    "                    gsm_data[\"disease\"] = sample_data.find(\"./Attributes/Attribute[@attribute_name='tissue']\").text\n",
    "            except:\n",
    "                gsm_data[\"disease\"] = gsm_data[\"disease\"] + \"\"\n",
    "\n",
    "            if (gsm_data[\"disease\"] == \"\" and \\\n",
    "                (\"tumor\" in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                \"tumour\" in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                \"normal\" in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower())\n",
    "               ):\n",
    "                    gsm_data[\"disease\"] = sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text\n",
    "            if ('organ' in str(subelem.attrib)): # no organ attributes at all\n",
    "                gsm_data[\"organ\"] = subelem.text\n",
    "            if ('tissue' in str(subelem.attrib) or \\\n",
    "                'brain region' in str(subelem.attrib)\n",
    "               ):\n",
    "                gsm_data[\"tissue\"] = subelem.text  # most organ info are from here # can have disease info\n",
    "            if ('cell_line' in str(subelem.attrib) or \\\n",
    "                'biological sample' in str(subelem.attrib) or \\\n",
    "                'neuronal subtype' in str(subelem.attrib) or \\\n",
    "                '\\'cell' in str(subelem.attrib)\n",
    "               ):\n",
    "                if ((not str(subelem.text).startswith(\"BRD\")) and (not str(subelem.text).startswith(\"MGH\"))): # these are gene names\n",
    "                    celltype_infos.append(str(subelem.attrib['attribute_name']) + \"-->\" + subelem.text)\n",
    "\n",
    "            if ('parental cell line' in str(subelem.attrib)):\n",
    "                celltype_infos.append(str(subelem.attrib['attribute_name']) + \"-->\" + subelem.text)\n",
    "                gsm_data[\"biomaterial\"] = \"cell line\"\n",
    "\n",
    "            if (\"mll-\" in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                \"bl6\" in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                \"hct\" in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                \"vcap\" in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                \"h9\" in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                \"cell line\" in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() or \\\n",
    "                (\"rim\" not in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower() and \\\n",
    "                 \"cyte\" in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower())\n",
    "               ):\n",
    "                celltype_infos.append(\"{source_name}\" + \"-->\" + str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text))\n",
    "                gsm_data[\"biomaterial\"] = \"cell line\"\n",
    "\n",
    "            if ('cell_line' in str(subelem.attrib)):\n",
    "                if (subelem.text != \"--\"):\n",
    "                    gsm_data[\"biomaterial\"] = \"cell line\"\n",
    "\n",
    "            if ('strain' in str(subelem.attrib) or 'mouse model' in str(subelem.attrib)):\n",
    "                gsm_data[\"strain\"] = subelem.text\n",
    "\n",
    "            if ('p14' in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower()):\n",
    "                gsm_data[\"strain\"] = str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text)\n",
    "\n",
    "            if ('gender' in str(subelem.attrib) or \\\n",
    "                '\\'sex\\'' in str(subelem.attrib)):\n",
    "                gsm_data[\"sex\"] = subelem.text\n",
    "\n",
    "            if (\"day\" in str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text).lower()):\n",
    "                gsm_data[\"age\"] = str(sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\").text)\n",
    "\n",
    "            if ('\\'age\\'' in str(subelem.attrib) or \\\n",
    "                'age_day' in str(subelem.attrib)):\n",
    "                gsm_data[\"age\"] = subelem.text\n",
    "\n",
    "            if ('\\'time' in str(subelem.attrib) or \\\n",
    "                '\\'differentiation' in str(subelem.attrib) or \\\n",
    "                'differentiation\\'' in str(subelem.attrib) or \\\n",
    "                'developmental stage' in str(subelem.attrib) or \\\n",
    "                ('erythropoiesis' in str(subelem.attrib) and 'no' not in str(subelem.attrib))):\n",
    "                if gsm_data[\"age\"] == \"\":\n",
    "                    gsm_data[\"age\"] = subelem.text\n",
    "            \n",
    "    gsm_data[\"celltype\"] = '|'.join(set(celltype_infos))\n",
    "\n",
    "    other = sample_data.find(\"./Attributes/Attribute[@attribute_name='source_name']\")\n",
    "    if other is not None:\n",
    "        gsm_data[\"other\"] = \"{source_name}\" + \"-->\" + str(other.text)\n",
    "    else:\n",
    "        gsm_data[\"other\"] = \"\"\n",
    "    \n",
    "    cell_type_df = cell_type_df.append(gsm_data, ignore_index=True)\n",
    "    \n",
    "    # indicate how many GSM are processed and save the table as the loop runs\n",
    "    num_of_gsm_done += 1\n",
    "    if (num_of_gsm_done % 100 == 0):\n",
    "            print(\"finished GSM number\", num_of_gsm_done)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7d515afe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>organism</th>\n",
       "      <th>biomaterial</th>\n",
       "      <th>disease</th>\n",
       "      <th>organ</th>\n",
       "      <th>tissue</th>\n",
       "      <th>celltype</th>\n",
       "      <th>strain</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>extdb_name</th>\n",
       "      <th>extdb_uuid</th>\n",
       "      <th>other</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>primary</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>primary</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>primary</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>primary</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>primary</td>\n",
       "      <td></td>\n",
       "      <td>embryo</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9164</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>primary</td>\n",
       "      <td></td>\n",
       "      <td>embryo</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9165</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>primary</td>\n",
       "      <td></td>\n",
       "      <td>embryo</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9166</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>primary</td>\n",
       "      <td></td>\n",
       "      <td>embryo</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9167</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>primary</td>\n",
       "      <td></td>\n",
       "      <td>embryo</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9168</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>primary</td>\n",
       "      <td></td>\n",
       "      <td>embryo</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9169 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     name organism biomaterial disease   organ tissue celltype strain sex age  \\\n",
       "0                      primary                                                  \n",
       "1                      primary                                                  \n",
       "2                      primary                                                  \n",
       "3                      primary                                                  \n",
       "4                      primary          embryo                                  \n",
       "...   ...      ...         ...     ...     ...    ...      ...    ...  ..  ..   \n",
       "9164                   primary          embryo                                  \n",
       "9165                   primary          embryo                                  \n",
       "9166                   primary          embryo                                  \n",
       "9167                   primary          embryo                                  \n",
       "9168                   primary          embryo                                  \n",
       "\n",
       "     extdb_name extdb_uuid other  \n",
       "0                                 \n",
       "1                                 \n",
       "2                                 \n",
       "3                                 \n",
       "4                                 \n",
       "...         ...        ...   ...  \n",
       "9164                              \n",
       "9165                              \n",
       "9166                              \n",
       "9167                              \n",
       "9168                              \n",
       "\n",
       "[9169 rows x 13 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_type_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c298725d",
   "metadata": {},
   "source": [
    "### Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d24f12b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output file:  GEO_Query_cell_type.2022_12_23_18_16\n"
     ]
    }
   ],
   "source": [
    "# setting the output filename\n",
    "today = date.today()\n",
    "date_str = today.strftime(\"%Y_%m_%d\")\n",
    "now = datetime.now()\n",
    "time_str = now.strftime(\"%H_%M\")\n",
    "output = \"GEO_Query_cell_type.{}_{}\".format(date_str, time_str)\n",
    "print(\"output file: \", output)\n",
    "cell_type_df.to_excel(output + \".xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5a70117",
   "metadata": {},
   "source": [
    "Sample ID: This is the unique identifier for each sample in the database. You can search for samples by their ID using this field.\n",
    "\n",
    "Organism: This field allows you to search for samples from a specific organism, such as human, mouse, or yeast.\n",
    "\n",
    "Tissue: This field allows you to search for samples based on their tissue type, such as liver, brain, or heart.\n",
    "\n",
    "Disease state: This field allows you to search for samples based on their disease state, such as cancer, diabetes, or autoimmune disease.\n",
    "\n",
    "Cell type: This field allows you to search for samples based on their cell type, such as fibroblasts, macrophages, or stem cells.\n",
    "\n",
    "Developmental stage: This field allows you to search for samples based on their developmental stage, such as embryonic, fetal, or adult.\n",
    "\n",
    "Experimental design: This field allows you to search for samples based on the type of experiment they were used in, such as microarray, RNA-seq, or ChIP-seq."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  },
  "vscode": {
   "interpreter": {
    "hash": "036b15a84498062e1e205bf5f2012dec27d9ce09ecb8b5e207c5080560abe931"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
